---
title: "multicrispr: Crispr/Cas9 library design"
author: A.M. Bhagwat, M. Looso
output: BiocStyle::html_document
bibliography: bibliography.bib
vignette: >
  %\VignetteIndexEntry{finding_flank_cas9s}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Crispr/Cas9

## Immune system turns into genome engineering

Since its initial discovery [@Ishino1987], and later understanding as a prokaryotic immune system [@Mojica2005; @Pourcel2005; @Biotin2005], Crispr/Cas9 has matured into a versatile molecular tool for genome engineering and analysis [@Jinek2012; @Gasiunas2012; @Cong2013]. As a molecular tool, it consists of two components: a 23-bp N~20~NGG **gRNA** (guide RNA) that targets the Cas9 enzyme to a particular genomic locus, and the **Cas9** enzyme itself, that performs some action at that locus. While the original S. pyogenes Cas9 helicase/endonuclease breaks double stranded DNA at the genomic locus of interest, the subsequently engineered (catalytically dead) **dCas9** used in many contemporary applications targets a genomic locus without cutting it, instead allowing for alternative activities through Cas9 gene fusion with an additional effector domain, such as KRAB-based **repression** [@Gilbert2013], VPR-based **activation**, [@Chavez2015], or APEX-based protein interactor **biotinylation** [@Myers2018].

## Designing gRNAs

The Cas9 enzyme works with an **N~20~NGG** gRNA, a sequence generic enough to allow for multiple alternative cas9 sequences for a genomic target site. Not all cas9 sequences are equally suited, though, and upstream bioinformatics work for proper gRNA design can greatly enhance the success of a cas9 experiment. Making sure an appropriate crispr/Cas9 sequence is chosen involves ensuring that the chosen gRNA **efficiently targets** the genomic locus of interest, but at the same time is **offtarget-free**, not binding any other genomic locus.

Targeting efficiency selection benefits greatly from the Cas9 knowledge gathered in recent years. Perhaps the most comprehensive work has been done by Doench et al, who investigated the effects of 1841 alternate gRNA sequences [@Doench2014] on targeting efficiency, later extended to more than 4000 alternative sequences [@Doench2016]. From these, a targeting efficiency prediction ruleset was extracted, initially using the initial SVM/logistic regression-based ruleset1 [@Doench2014], later improved into the boosted regression tree-based ruleset2 [@Doench2016].

## From plasmids to libraries

Initially, Crispr/Cas9 was used to target a single genomic locus. Later advances alowed for the targeting of multiple loci, through gRNA multiplexing on a single plasmid. Subsequent advances now enable mass targeting of thousands of genomic loci, through a gRNA library, with many different single gRNA clones. This mass targeting off course increases algorithmic/computational demands for gRNA design, since thousands of genomic target sites allow for many thousands of potential crispr sites, each of which require (mismatch-aware) offtarget analysis as well as targeting efficiency predictions to be performed.

## Designing Crispr/Cas9 libraries: the multicrispr package

`multicrispr` has been developed to serve these needs in mass gRNA design. It provides intuitive functions to (1) read a bedfile with many genomic ranges into a `GRanges` object, (2) extend or flank these when required, (3) visualize the effects of such operations, (4) examine offtarget effects and filter for offtarget-free cas9 sequences, and (5) compute the predicted targeting efficiency and filter for efficient cas9 sequences.

Care has been taken to make `multicrispr` intuitive and efficient, much of which is achieved by aligning its functionality with the existing R/Bioconductor functionality, re-using the dedicated classes `GRanges` (genomic loci), `Biostrings` (genomic sequences) and `BSgenome` (full genomes), each with dedicated, efficient, c-rooted methods (e.g. for mismatch-aware offtarget analysis), and by performing operations in a data.table context when appropriate, enabling c-level loops. Intuition is served by having most functions output both graphical and textual summaries, allowing immediate inspection and anomaly detection. In the following sections, we discuss each of these steps in more detail.

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment  = "    #", 
  cache    = TRUE
)

```

# The multicrispr workflow

## Define target ranges

Target ranges can be defined as **Entrez** or **Ensembl** gene identifiers (`genes_to_granges`). Entrez geneids are mapped to ranges using (UCSC-based) TxDb gene models, while Ensembl geneids are mapped to ranges using (Ensembl-based) EnsDb gene models. TxDb and EnsDb objects can be accessed through Bioconductor's AnnotationHub data repository, which provides gene models for all organisms with a completed sequence:

```{r read}
# Connect to AnnotationHub and specify data dir
    require(multicrispr)
    ah <- AnnotationHub::AnnotationHub()
    dir <- system.file('extdata',  package='multicrispr')

# Entrez -> GRanges
    AnnotationHub::query(ah, c('TxDb', 'Mus musculus', '2019'))
    db <- ah[['AH70594']]
    genefile <- paste0(dir, '/SRF.entrez')
    gr <- genefile_to_granges(genefile, db)

# Ensembl -> GRanges 
    AnnotationHub::query(ah, c('EnsDb', 'Mus musculus', '2019'))
    db <- ah[['AH75036']]
    genefile <- paste0(dir, '/SRF.ensembl');  
    gr <- genefile_to_granges(genefile, db)
```

Target ranges can also be defined through a BED file with genomic coordinates:

```{r read}
   # BED file
    bedfile <- system.file('extdata/SRF.bed', package = 'multicrispr')
    gr <- bed_to_granges(bedfile, genome = 'mm10')      # UCSC genome notation
```

## Extend or Flank

Target ranges may require extension (to allow enclosure of a 23 bp crispr sequence) or flanking (to target enhancers), which is achieved through the functions `extend`, `left_flank`, and `right_flank`. 

```{r leftflank}
targets <-  left_flank(gr, -100,  -1,  plot = FALSE)
targets <- right_flank(gr,    1, 100,  plot = FALSE)
targets <-      extend(gr,  -22,  22,  plot = FALSE)
```
## Find cas9 sites

Next, N~20~NGG$ Cas9 sequences are searched in the target ranges

```{r findcas9s, eval=FALSE}
cas9s <- find_cas9s(targets)
```


## Filter for no offtargets

```{r findcas9s, eval=FALSE}
cas9s <- filter_no_offtargets(cas9s, targets)
```

## Score cas9 sites

Doench et al. (2014)'s "ruleset1" for scoring cas9 sequences is readily available:

```{r score_rs1, eval = FALSE}
cas9ranges$score_rs1 <- score_cas9ranges(cas9ranges, ruleset = 1)
```

The improved "ruleset2" (Doench, 2016) is also accessible, but requires for the python module  [azimuth](https://github.com/MicrosoftResearch/Azimuth]) to be installed and accessible as a python binary, virtual environment, or conda environment. The code chunk below shows how to score using ruleset 2 if azimuth was installed in the condaenv "default". It also shows the common operation of filtering for ruleset2 score values above 0.4.


```{r score_rs2, eval = FALSE}
python_available <- assertive.reflection::is_unix() & 
                    reticulate::py_module_available('azimuth')
if (python_available){
  cas9s$score_rs2 <- score_cas9s(cas9s, ruleset = 2, condaenv = 'default')
  
  message('\tFilter for scores > 0.4')
  cas9s %<>% extract(cas9s$score > 0.4)
  cmessage('\t\t%d cas9 seqs across %d ranges', 
            length(unique(cas9s$seqs)), 
            length(cas9s))
}
```

# References

Doench et al. (2014). Rational design of highly active sgRNAs for CRISPR-Cas9-mediated gene 
inactivation. Nature Biotechnology, 32(12), doi:10.1038/nbt.3026

Doench et al. (2016). Optimized sgRNA design to maximize activity and minimize off-target effects of CRISPR-Cas9. Nature Biotechnology, 34(7), doi:10.1038/nbt.3437

Ishono, Shinagawa, Makino, Amemura, Nakata (1987). Nucleotide sequence of the iap gene, responsible for alkaline phosphatase isozyme conversion in Escherichia coli, and identification of the gene product. J. Bacteriology 169:5429-5433. doi:10.1128/jb.169.12.5429-5433.1987.


