<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Add efficiency scores and filter — add_efficiency • multicrispr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Add efficiency scores and filter — add_efficiency" />
<meta property="og:description" content="Add Doench2014 or Doench2016 efficiency scores and filter on them" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">multicrispr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/designing_grna_libraries.html">multicrispr: gRNA design for Crispr/Cas9 and Prime Editing</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Add efficiency scores and filter</h1>
    
    <div class="hidden name"><code>add_efficiency.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Add Doench2014 or Doench2016 efficiency scores and filter on them</p>
    </div>

    <pre class="usage"><span class='fu'>add_efficiency</span>(
  <span class='no'>spacers</span>,
  <span class='no'>bsgenome</span>,
  <span class='kw'>method</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"Doench2014"</span>, <span class='st'>"Doench2016"</span>)[<span class='fl'>1</span>],
  <span class='kw'>chunksize</span> <span class='kw'>=</span> <span class='fl'>10000</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>plot</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>alpha_var</span> <span class='kw'>=</span> <span class='fu'>default_alpha_var</span>(<span class='no'>spacers</span>)
)

<span class='fu'>filter_efficient</span>(
  <span class='no'>spacers</span>,
  <span class='no'>bsgenome</span>,
  <span class='kw'>method</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"Doench2014"</span>, <span class='st'>"Doench2016"</span>)[<span class='fl'>1</span>],
  <span class='no'>cutoff</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>plot</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>alpha_var</span> <span class='kw'>=</span> <span class='fu'>default_alpha_var</span>(<span class='no'>spacers</span>)
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>spacers</th>
      <td><p><code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges-class</a></code>: spacers</p></td>
    </tr>
    <tr>
      <th>bsgenome</th>
      <td><p><code><a href='https://rdrr.io/pkg/BSgenome/man/BSgenome-class.html'>BSgenome-class</a></code></p></td>
    </tr>
    <tr>
      <th>method</th>
      <td><p>'Doench2014' (default) or 'Doench2016'
(requires non-NULL argument python, virtualenv, or condaenv)</p></td>
    </tr>
    <tr>
      <th>chunksize</th>
      <td><p>Doench2016 is executed in chunks of chunksize</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>TRUE (default) or FALSE</p></td>
    </tr>
    <tr>
      <th>plot</th>
      <td><p>TRUE (default) or FALSE</p></td>
    </tr>
    <tr>
      <th>alpha_var</th>
      <td><p>NULL or string: var mapped to alpha in plot</p></td>
    </tr>
    <tr>
      <th>cutoff</th>
      <td><p>value to filter on</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>numeric vector</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><code>add_efficiency</code> adds efficiency scores
<code>filter_efficiency</code> adds efficiency scores and filters on them</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Doench 2014, Rational design of highly active sgRNAs for 
CRISPR-Cas9-mediated gene inactivation. Nature Biotechnology,
doi: 10.1038/nbt.3026</p>
<p>Doench 2016, Optimized sgRNA design to maximize activity and minimize 
off-target effects of CRISPR-Cas9. Nature Biotechnology, 
doi: 10.1038/nbt.3437</p>
<p>Python module azimuth: github/MicrosoftResearch/azimuth</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='co'># Install azimuth </span>
<span class='co'>#----------------</span>
    <span class='co'>## With reticulate</span>
    <span class='co'># require(reticulate)</span>
    <span class='co'># conda_create('azienv', c('python=2.7'))</span>
    <span class='co'># use_condaenv('azienv')</span>
    <span class='co'># py_install(c('azimuth', 'scikit-learn==0.17.1'), 'azienv', pip = TRUE)</span>

    <span class='co'>## Directly</span>
    <span class='co'># conda create --name azienv python=2.7</span>
    <span class='co'># conda activate azienv</span>
    <span class='co'># pip install azimuth</span>
    <span class='co'># pip install scikit-learn==0.17.1</span>

<span class='co'># PE example</span>
<span class='co'>#-----------</span>
    <span class='fu'><a href='https://rdrr.io/r/base/library.html'>require</a></span>(<span class='no'>magrittr</span>)
    <span class='no'>bsgenome</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>BSgenome.Hsapiens.UCSC.hg38</span><span class='kw ns'>::</span><span class='no'><a href='https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg38/man/package.html'>BSgenome.Hsapiens.UCSC.hg38</a></span>
    <span class='no'>targets</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='char_to_granges.html'>char_to_granges</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='kw'>PRNP</span> <span class='kw'>=</span> <span class='st'>'chr20:4699600:+'</span>,             <span class='co'># snp</span>
                                 <span class='kw'>HBB</span>  <span class='kw'>=</span> <span class='st'>'chr11:5227002:-'</span>,             <span class='co'># snp</span>
                                 <span class='kw'>HEXA</span> <span class='kw'>=</span> <span class='st'>'chr15:72346580-72346583:-'</span>,   <span class='co'># del</span>
                                 <span class='kw'>CFTR</span> <span class='kw'>=</span> <span class='st'>'chr7:117559593-117559595:+'</span>), <span class='co'># ins</span>
                               <span class='no'>bsgenome</span>)
    <span class='no'>spacers</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='find_pe_spacers.html'>find_pe_spacers</a></span>(<span class='no'>targets</span>, <span class='no'>bsgenome</span>)</div><div class='img'><img src='add_efficiency-1.png' alt='' width='700' height='433' /></div><div class='input'>   <span class='co'>#spacers &lt;- find_spacers(extend_for_pe(gr), bsgenome, complement = FALSE)</span>
    (<span class='no'>spacers</span> <span class='kw'>%&lt;&gt;%</span> <span class='fu'>add_efficiency</span>(<span class='no'>bsgenome</span>, <span class='st'>'Doench2014'</span>))</div><div class='output co'>#&gt; <span class='message'>	Score crispr spacers</span></div><div class='output co'>#&gt; <span class='message'>		Add (4-23-3) contextseqs</span></div><div class='output co'>#&gt; <span class='message'>		Score contextseqs with Doench2014</span></div><div class='img'><img src='add_efficiency-2.png' alt='' width='700' height='433' /></div><div class='output co'>#&gt; GRanges object with 10 ranges and 11 metadata columns:
#&gt;          seqnames              ranges strand |                  crisprcontext
#&gt;             &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |                    &lt;character&gt;
#&gt;   CFTR_1     chr7 117559575-117559594      + | CACCATTAAAGAAAATATCATCTTTGGTGT
#&gt;   CFTR_2     chr7 117559606-117559625      - | CGCTTCTGTATCTATATTCATCATAGGAAA
#&gt;    HBB_1    chr11     5227003-5227022      - | ACACCATGGTGCATCTGACTCCTGAGGAGA
#&gt;    HBB_2    chr11     5226984-5227003      + | GGCAGTAACGGCAGACTTCTCCTCAGGAGT
#&gt;   HEXA_1    chr15   72346551-72346570      + | ACTATGTAGAAATCCTTCCAGTCAGGGCCA
#&gt;   HEXA_2    chr15   72346558-72346577      + | AGAAATCCTTCCAGTCAGGGCCATAGGATA
#&gt;   PRNP_1    chr20     4699568-4699587      + | CTGCAGCAGCTGGGGCAGTGGTGGGGGGCC
#&gt;   PRNP_2    chr20     4699569-4699588      + | TGCAGCAGCTGGGGCAGTGGTGGGGGGCCT
#&gt;   PRNP_3    chr20     4699575-4699594      + | AGCTGGGGCAGTGGTGGGGGGCCTTGGCGG
#&gt;   PRNP_4    chr20     4699578-4699597      + | TGGGGCAGTGGTGGGGGGCCTTGGCGGCTA
#&gt;           targetname targetstart targetend  crisprname         crisprspacer
#&gt;          &lt;character&gt;   &lt;integer&gt; &lt;integer&gt; &lt;character&gt;          &lt;character&gt;
#&gt;   CFTR_1        CFTR   117559593 117559595      CFTR_1 ATTAAAGAAAATATCATCTT
#&gt;   CFTR_2        CFTR   117559593 117559595      CFTR_2 TCTGTATCTATATTCATCAT
#&gt;    HBB_1         HBB     5227002   5227002       HBB_1 CATGGTGCATCTGACTCCTG
#&gt;    HBB_2         HBB     5227002   5227002       HBB_2 GTAACGGCAGACTTCTCCTC
#&gt;   HEXA_1        HEXA    72346580  72346583      HEXA_1 TGTAGAAATCCTTCCAGTCA
#&gt;   HEXA_2        HEXA    72346580  72346583      HEXA_2 ATCCTTCCAGTCAGGGCCAT
#&gt;   PRNP_1        PRNP     4699600   4699600      PRNP_1 AGCAGCTGGGGCAGTGGTGG
#&gt;   PRNP_2        PRNP     4699600   4699600      PRNP_2 GCAGCTGGGGCAGTGGTGGG
#&gt;   PRNP_3        PRNP     4699600   4699600      PRNP_3 GGGGCAGTGGTGGGGGGCCT
#&gt;   PRNP_4        PRNP     4699600   4699600      PRNP_4 GCAGTGGTGGGGGGCCTTGG
#&gt;            crisprpam        primer    revtranscript
#&gt;          &lt;character&gt;   &lt;character&gt;      &lt;character&gt;
#&gt;   CFTR_1         TGG AAGAAAATATCAT CTTTGGTGTTTCCTAT
#&gt;   CFTR_2         AGG TATCTATATTCAT CATAGGAAACACCAAA
#&gt;    HBB_1         AGG GTGCATCTGACTC CTGAGGAGAAGTCTGC
#&gt;    HBB_2         AGG CGGCAGACTTCTC CTCAGGAGTCAGATGC
#&gt;   HEXA_1         GGG GAAATCCTTCCAG TCAGGGCCATAGGATA
#&gt;   HEXA_2         AGG TTCCAGTCAGGGC CATAGGATATACGGTT
#&gt;   PRNP_1         GGG GCTGGGGCAGTGG TGGGGGGCCTTGGCGG
#&gt;   PRNP_2         GGG CTGGGGCAGTGGT GGGGGGCCTTGGCGGC
#&gt;   PRNP_3         TGG CAGTGGTGGGGGG CCTTGGCGGCTACATG
#&gt;   PRNP_4         CGG TGGTGGGGGGCCT TGGCGGCTACATGCTG
#&gt;                              extension          Doench2014
#&gt;                            &lt;character&gt;           &lt;numeric&gt;
#&gt;   CFTR_1 ATAGGAAACACCAAAGATGATATTTTCTT  0.0339998970537971
#&gt;   CFTR_2 TTTGGTGTTTCCTATGATGAATATAGATA   0.223370560176775
#&gt;    HBB_1 GCAGACTTCTCCTCAGGAGTCAGATGCAC    0.13111422370078
#&gt;    HBB_2 GCATCTGACTCCTGAGGAGAAGTCTGCCG   0.413980645125447
#&gt;   HEXA_1 TATCCTATGGCCCTGACTGGAAGGATTTC   0.107283614825074
#&gt;   HEXA_2 AACCGTATATCCTATGGCCCTGACTGGAA  0.0394252780338432
#&gt;   PRNP_1 CCGCCAAGGCCCCCCACCACTGCCCCAGC 0.00994899162104746
#&gt;   PRNP_2 GCCGCCAAGGCCCCCCACCACTGCCCCAG  0.0449662673405108
#&gt;   PRNP_3 CATGTAGCCGCCAAGGCCCCCCACCACTG 0.00996253852673941
#&gt;   PRNP_4 CAGCATGTAGCCGCCAAGGCCCCCCACCA    0.20935862741641
#&gt;   -------
#&gt;   seqinfo: 455 sequences (1 circular) from hg38 genome</div><div class='input'>    <span class='co'># reticulate::use_condaenv('azienv')</span>
    <span class='co'># reticulate::import('azimuth')</span>
    <span class='co'># spacers %&lt;&gt;% add_efficiency(bsgenome, 'Doench2016')</span>
    <span class='co'># filter_efficient(spacers, bsgenome, 'Doench2016', 0.4)</span>

<span class='co'># TFBS example</span>
<span class='co'>#-------------</span>
    <span class='no'>bedfile</span>  <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'extdata/SRF.bed'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'multicrispr'</span>)
    <span class='no'>bsgenome</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>BSgenome.Mmusculus.UCSC.mm10</span><span class='kw ns'>::</span><span class='no'><a href='https://rdrr.io/pkg/BSgenome.Mmusculus.UCSC.mm10/man/package.html'>BSgenome.Mmusculus.UCSC.mm10</a></span>
    <span class='no'>targets</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='up_flank.html'>extend</a></span>(<span class='fu'><a href='bed_to_granges.html'>bed_to_granges</a></span>(<span class='no'>bedfile</span>, <span class='st'>'mm10'</span>))</div><div class='output co'>#&gt; <span class='message'>	Read SRF.bed into GRanges</span></div><div class='output co'>#&gt; <span class='message'>		1974 ranges on 21 chromosomes</span></div><div class='img'><img src='add_efficiency-3.png' alt='' width='700' height='433' /></div><div class='input'>    <span class='no'>spacers</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='find_spacers.html'>find_spacers</a></span>(<span class='no'>targets</span>, <span class='no'>bsgenome</span>)</div><div class='output co'>#&gt; <span class='message'>		3948 ranges after adding inverse strands</span></div><div class='img'><img src='add_efficiency-4.png' alt='' width='700' height='433' /></div><div class='input'>    # reticulate::use_condaenv('azienv')
    # reticulate::import('azienv')
    # (spacers %<>% add_specificity(targets, bsgenome))
    # (spacers %>% add_efficiency(bsgenome, 'Doench2014'))
    # (spacers %>% add_efficiency(bsgenome, 'Doench2016'))
    #  spacers %>% filter_efficient(bsgenome, 'Doench2016', 0.4)
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#references">References</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Aditya Bhagwat.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


