<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Find crispr spacers in targetranges — find_spacers • multicrispr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Find crispr spacers in targetranges — find_spacers" />
<meta property="og:description" content="Find crispr spacers in targetranges" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">multicrispr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/designing_grna_libraries.html">Using multicrispr for gRNA design</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Find crispr spacers in targetranges</h1>
    
    <div class="hidden name"><code>find_spacers.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Find crispr spacers in targetranges</p>
    </div>

    <pre class="usage"><span class='fu'>find_spacers</span>(
  <span class='no'>gr</span>,
  <span class='no'>bsgenome</span>,
  <span class='kw'>spacer</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/strrep.html'>strrep</a></span>(<span class='st'>"N"</span>, <span class='fl'>20</span>),
  <span class='kw'>pam</span> <span class='kw'>=</span> <span class='st'>"NGG"</span>,
  <span class='kw'>complement</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>plot</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>gr</th>
      <td><p><code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges-class</a></code></p></td>
    </tr>
    <tr>
      <th>bsgenome</th>
      <td><p><code><a href='https://rdrr.io/pkg/BSgenome/man/BSgenome-class.html'>BSgenome-class</a></code></p></td>
    </tr>
    <tr>
      <th>spacer</th>
      <td><p>string: spacer pattern in extended IUPAC alphabet</p></td>
    </tr>
    <tr>
      <th>pam</th>
      <td><p>string: pam pattern in extended IUPAC alphabet</p></td>
    </tr>
    <tr>
      <th>complement</th>
      <td><p>TRUE (default) or FALSE: also search in compl ranges?</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>TRUE (default) or FALSE</p></td>
    </tr>
    <tr>
      <th>plot</th>
      <td><p>TRUE (default) or FALSE</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges-class</a></code></p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='find_pe_spacers.html'>find_pe_spacers</a></code> to find prime editing spacers</p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># PE example</span>
<span class='co'>#-----------</span>
    <span class='fu'><a href='https://rdrr.io/r/base/library.html'>require</a></span>(<span class='no'>magrittr</span>)
    <span class='no'>bsgenome</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>BSgenome.Hsapiens.UCSC.hg38</span><span class='kw ns'>::</span><span class='no'><a href='https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg38/man/package.html'>BSgenome.Hsapiens.UCSC.hg38</a></span>
    <span class='no'>gr</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='char_to_granges.html'>char_to_granges</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='kw'>PRNP</span> <span class='kw'>=</span> <span class='st'>'chr20:4699600:+'</span>,             <span class='co'># snp</span>
                            <span class='kw'>HBB</span>  <span class='kw'>=</span> <span class='st'>'chr11:5227002:-'</span>,             <span class='co'># snp</span>
                            <span class='kw'>HEXA</span> <span class='kw'>=</span> <span class='st'>'chr15:72346580-72346583:-'</span>,   <span class='co'># del</span>
                            <span class='kw'>CFTR</span> <span class='kw'>=</span> <span class='st'>'chr7:117559593-117559595:+'</span>), <span class='co'># ins</span>
                          <span class='no'>bsgenome</span>)
    <span class='fu'><a href='plot_intervals.html'>plot_intervals</a></span>(<span class='no'>gr</span>)</div><div class='img'><img src='find_spacers-1.png' alt='' width='700' height='433' /></div><div class='input'>    <span class='fu'><a href='find_pe_spacers.html'>find_pe_spacers</a></span>(<span class='no'>gr</span>, <span class='no'>bsgenome</span>)</div><div class='img'><img src='find_spacers-2.png' alt='' width='700' height='433' /></div><div class='output co'>#&gt; GRanges object with 10 ranges and 9 metadata columns:
#&gt;          seqnames              ranges strand |  targetname targetstart
#&gt;             &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;   &lt;integer&gt;
#&gt;   CFTR_1     chr7 117559575-117559594      + |        CFTR   117559593
#&gt;   CFTR_2     chr7 117559606-117559625      - |        CFTR   117559593
#&gt;    HBB_1    chr11     5227003-5227022      - |         HBB     5227002
#&gt;    HBB_2    chr11     5226984-5227003      + |         HBB     5227002
#&gt;   HEXA_1    chr15   72346551-72346570      + |        HEXA    72346580
#&gt;   HEXA_2    chr15   72346558-72346577      + |        HEXA    72346580
#&gt;   PRNP_1    chr20     4699568-4699587      + |        PRNP     4699600
#&gt;   PRNP_2    chr20     4699569-4699588      + |        PRNP     4699600
#&gt;   PRNP_3    chr20     4699575-4699594      + |        PRNP     4699600
#&gt;   PRNP_4    chr20     4699578-4699597      + |        PRNP     4699600
#&gt;          targetend  crisprname         crisprspacer   crisprpam        primer
#&gt;          &lt;integer&gt; &lt;character&gt;          &lt;character&gt; &lt;character&gt;   &lt;character&gt;
#&gt;   CFTR_1 117559595      CFTR_1 ATTAAAGAAAATATCATCTT         TGG AAGAAAATATCAT
#&gt;   CFTR_2 117559595      CFTR_2 TCTGTATCTATATTCATCAT         AGG TATCTATATTCAT
#&gt;    HBB_1   5227002       HBB_1 CATGGTGCATCTGACTCCTG         AGG GTGCATCTGACTC
#&gt;    HBB_2   5227002       HBB_2 GTAACGGCAGACTTCTCCTC         AGG CGGCAGACTTCTC
#&gt;   HEXA_1  72346583      HEXA_1 TGTAGAAATCCTTCCAGTCA         GGG GAAATCCTTCCAG
#&gt;   HEXA_2  72346583      HEXA_2 ATCCTTCCAGTCAGGGCCAT         AGG TTCCAGTCAGGGC
#&gt;   PRNP_1   4699600      PRNP_1 AGCAGCTGGGGCAGTGGTGG         GGG GCTGGGGCAGTGG
#&gt;   PRNP_2   4699600      PRNP_2 GCAGCTGGGGCAGTGGTGGG         GGG CTGGGGCAGTGGT
#&gt;   PRNP_3   4699600      PRNP_3 GGGGCAGTGGTGGGGGGCCT         TGG CAGTGGTGGGGGG
#&gt;   PRNP_4   4699600      PRNP_4 GCAGTGGTGGGGGGCCTTGG         CGG TGGTGGGGGGCCT
#&gt;             revtranscript                     extension
#&gt;               &lt;character&gt;                   &lt;character&gt;
#&gt;   CFTR_1 CTTTGGTGTTTCCTAT ATAGGAAACACCAAAGATGATATTTTCTT
#&gt;   CFTR_2 CATAGGAAACACCAAA TTTGGTGTTTCCTATGATGAATATAGATA
#&gt;    HBB_1 CTGAGGAGAAGTCTGC GCAGACTTCTCCTCAGGAGTCAGATGCAC
#&gt;    HBB_2 CTCAGGAGTCAGATGC GCATCTGACTCCTGAGGAGAAGTCTGCCG
#&gt;   HEXA_1 TCAGGGCCATAGGATA TATCCTATGGCCCTGACTGGAAGGATTTC
#&gt;   HEXA_2 CATAGGATATACGGTT AACCGTATATCCTATGGCCCTGACTGGAA
#&gt;   PRNP_1 TGGGGGGCCTTGGCGG CCGCCAAGGCCCCCCACCACTGCCCCAGC
#&gt;   PRNP_2 GGGGGGCCTTGGCGGC GCCGCCAAGGCCCCCCACCACTGCCCCAG
#&gt;   PRNP_3 CCTTGGCGGCTACATG CATGTAGCCGCCAAGGCCCCCCACCACTG
#&gt;   PRNP_4 TGGCGGCTACATGCTG CAGCATGTAGCCGCCAAGGCCCCCCACCA
#&gt;   -------
#&gt;   seqinfo: 455 sequences (1 circular) from hg38 genome</div><div class='input'>    <span class='fu'>find_spacers</span>(<span class='fu'><a href='extend_for_pe.html'>extend_for_pe</a></span>(<span class='no'>gr</span>), <span class='no'>bsgenome</span>, <span class='kw'>complement</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</div><div class='img'><img src='find_spacers-3.png' alt='' width='700' height='433' /></div><div class='output co'>#&gt; GRanges object with 10 ranges and 6 metadata columns:
#&gt;            seqnames              ranges strand |  targetname targetstart
#&gt;               &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;   &lt;integer&gt;
#&gt;     CFTR_f     chr7 117559575-117559594      + |        CFTR   117559593
#&gt;     CFTR_r     chr7 117559606-117559625      - |        CFTR   117559593
#&gt;      HBB_f    chr11     5226984-5227003      + |         HBB     5227002
#&gt;      HBB_r    chr11     5227003-5227022      - |         HBB     5227002
#&gt;   HEXA_f_1    chr15   72346551-72346570      + |        HEXA    72346580
#&gt;   HEXA_f_2    chr15   72346558-72346577      + |        HEXA    72346580
#&gt;   PRNP_f_1    chr20     4699568-4699587      + |        PRNP     4699600
#&gt;   PRNP_f_2    chr20     4699569-4699588      + |        PRNP     4699600
#&gt;   PRNP_f_3    chr20     4699575-4699594      + |        PRNP     4699600
#&gt;   PRNP_f_4    chr20     4699578-4699597      + |        PRNP     4699600
#&gt;            targetend  crisprname         crisprspacer   crisprpam
#&gt;            &lt;integer&gt; &lt;character&gt;          &lt;character&gt; &lt;character&gt;
#&gt;     CFTR_f 117559595      CFTR_f ATTAAAGAAAATATCATCTT         TGG
#&gt;     CFTR_r 117559595      CFTR_r TCTGTATCTATATTCATCAT         AGG
#&gt;      HBB_f   5227002       HBB_f GTAACGGCAGACTTCTCCTC         AGG
#&gt;      HBB_r   5227002       HBB_r CATGGTGCATCTGACTCCTG         AGG
#&gt;   HEXA_f_1  72346583    HEXA_f_1 TGTAGAAATCCTTCCAGTCA         GGG
#&gt;   HEXA_f_2  72346583    HEXA_f_2 ATCCTTCCAGTCAGGGCCAT         AGG
#&gt;   PRNP_f_1   4699600    PRNP_f_1 AGCAGCTGGGGCAGTGGTGG         GGG
#&gt;   PRNP_f_2   4699600    PRNP_f_2 GCAGCTGGGGCAGTGGTGGG         GGG
#&gt;   PRNP_f_3   4699600    PRNP_f_3 GGGGCAGTGGTGGGGGGCCT         TGG
#&gt;   PRNP_f_4   4699600    PRNP_f_4 GCAGTGGTGGGGGGCCTTGG         CGG
#&gt;   -------
#&gt;   seqinfo: 455 sequences (1 circular) from hg38 genome</div><div class='input'>          <span class='co'># complement = FALSE because extend_for_pe  already </span>
          <span class='co'># adds  reverse complements and does so in a strand-specific </span>
          <span class='co'># manner</span>

<span class='co'># TFBS example</span>
<span class='co'>#-------------</span>
    <span class='no'>bsgenome</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>BSgenome.Mmusculus.UCSC.mm10</span><span class='kw ns'>::</span><span class='no'><a href='https://rdrr.io/pkg/BSgenome.Mmusculus.UCSC.mm10/man/package.html'>BSgenome.Mmusculus.UCSC.mm10</a></span>
    <span class='no'>bedfile</span>  <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'extdata/SRF.bed'</span>, <span class='kw'>package</span><span class='kw'>=</span><span class='st'>'multicrispr'</span>)
    <span class='no'>gr</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='bed_to_granges.html'>bed_to_granges</a></span>(<span class='no'>bedfile</span>, <span class='st'>'mm10'</span>) <span class='kw'>%&gt;%</span> <span class='fu'><a href='up_flank.html'>extend</a></span>()</div><div class='output co'>#&gt; <span class='message'>	Read SRF.bed into GRanges</span></div><div class='output co'>#&gt; <span class='message'>		1974 ranges on 21 chromosomes</span></div><div class='img'><img src='find_spacers-4.png' alt='' width='700' height='433' /></div><div class='input'>    <span class='fu'>find_spacers</span>(<span class='no'>gr</span>, <span class='no'>bsgenome</span>)</div><div class='output co'>#&gt; <span class='message'>		3948 ranges after adding inverse strands</span></div><div class='img'><img src='find_spacers-5.png' alt='' width='700' height='433' /></div><div class='output co'>#&gt; GRanges object with 10159 ranges and 8 metadata columns:
#&gt;             seqnames            ranges strand |         name     score
#&gt;                &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;character&gt; &lt;numeric&gt;
#&gt;   T0001_r_5     chr1   4712619-4712638      - | SRF_MA0083.3  10.49542
#&gt;   T0001_r_4     chr1   4712620-4712639      - | SRF_MA0083.3  10.49542
#&gt;   T0001_r_3     chr1   4712627-4712646      - | SRF_MA0083.3  10.49542
#&gt;   T0001_r_2     chr1   4712633-4712652      - | SRF_MA0083.3  10.49542
#&gt;   T0001_r_1     chr1   4712634-4712653      - | SRF_MA0083.3  10.49542
#&gt;         ...      ...               ...    ... .          ...       ...
#&gt;   T1974_r_3     chrY 89126492-89126511      - | SRF_MA0083.3   4.54393
#&gt;   T1974_f_3     chrY 89126496-89126515      + | SRF_MA0083.3   4.54393
#&gt;   T1974_r_2     chrY 89126499-89126518      - | SRF_MA0083.3   4.54393
#&gt;   T1974_r_1     chrY 89126500-89126519      - | SRF_MA0083.3   4.54393
#&gt;   T1974_f_4     chrY 89126501-89126520      + | SRF_MA0083.3   4.54393
#&gt;              targetname targetstart targetend  crisprname         crisprspacer
#&gt;             &lt;character&gt;   &lt;integer&gt; &lt;integer&gt; &lt;character&gt;          &lt;character&gt;
#&gt;   T0001_r_5       T0001     4712628   4712643   T0001_r_5 ATATAAGGGCATTGGAAGAA
#&gt;   T0001_r_4       T0001     4712628   4712643   T0001_r_4 AATATAAGGGCATTGGAAGA
#&gt;   T0001_r_3       T0001     4712628   4712643   T0001_r_3 TGGAGACAATATAAGGGCAT
#&gt;   T0001_r_2       T0001     4712628   4712643   T0001_r_2 TTCTGCTGGAGACAATATAA
#&gt;   T0001_r_1       T0001     4712628   4712643   T0001_r_1 CTTCTGCTGGAGACAATATA
#&gt;         ...         ...         ...       ...         ...                  ...
#&gt;   T1974_r_3       T1974    89126494  89126509   T1974_r_3 TTTGCCCATACTAGGGAGAG
#&gt;   T1974_f_3       T1974    89126494  89126509   T1974_f_3 CCCTAGTATGGGCAAATATA
#&gt;   T1974_r_2       T1974    89126494  89126509   T1974_r_2 CCATATATTTGCCCATACTA
#&gt;   T1974_r_1       T1974    89126494  89126509   T1974_r_1 TCCATATATTTGCCCATACT
#&gt;   T1974_f_4       T1974    89126494  89126509   T1974_f_4 GTATGGGCAAATATATGGAA
#&gt;               crisprpam
#&gt;             &lt;character&gt;
#&gt;   T0001_r_5         GGG
#&gt;   T0001_r_4         AGG
#&gt;   T0001_r_3         TGG
#&gt;   T0001_r_2         GGG
#&gt;   T0001_r_1         AGG
#&gt;         ...         ...
#&gt;   T1974_r_3         CGG
#&gt;   T1974_f_3         TGG
#&gt;   T1974_r_2         GGG
#&gt;   T1974_r_1         AGG
#&gt;   T1974_f_4         TGG
#&gt;   -------
#&gt;   seqinfo: 66 sequences (1 circular) from mm10 genome</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#see-also">See also</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Aditya Bhagwat.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


