<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Find prime editing spacers — find_primespacers • multicrispr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Find prime editing spacers — find_primespacers" />
<meta property="og:description" content="Find prime editing spacers around target ranges" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">multicrispr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.24</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/crispr_grna_design.html">Crispr/Cas9 gRNA design</a>
    </li>
    <li>
      <a href="../articles/genome_arithmetics.html">Genome arithmetics</a>
    </li>
    <li>
      <a href="../articles/prime_editing.html">Prime Editing</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Find prime editing spacers</h1>
    
    <div class="hidden name"><code>find_primespacers.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Find prime editing spacers around target ranges</p>
    </div>

    <pre class="usage"><span class='fu'>find_primespacers</span>(
  <span class='no'>gr</span>,
  <span class='no'>bsgenome</span>,
  <span class='kw'>edits</span> <span class='kw'>=</span> <span class='fu'>get_plus_seq</span>(<span class='no'>bsgenome</span>, <span class='no'>gr</span>),
  <span class='kw'>nprimer</span> <span class='kw'>=</span> <span class='fl'>13</span>,
  <span class='kw'>nrt</span> <span class='kw'>=</span> <span class='fl'>16</span>,
  <span class='kw'>ontargetmethod</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"Doench2014"</span>, <span class='st'>"Doench2016"</span>)[<span class='fl'>1</span>],
  <span class='kw'>offtargetmethod</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"bowtie"</span>, <span class='st'>"pdict"</span>)[<span class='fl'>1</span>],
  <span class='kw'>mismatches</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>nickmatches</span> <span class='kw'>=</span> <span class='fl'>2</span>,
  <span class='kw'>indexedgenomesdir</span> <span class='kw'>=</span> <span class='no'>INDEXEDGENOMESDIR</span>,
  <span class='kw'>outdir</span> <span class='kw'>=</span> <span class='no'>OUTDIR</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>plot</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='no'>...</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>gr</th>
      <td><p><code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges-class</a></code></p></td>
    </tr>
    <tr>
      <th>bsgenome</th>
      <td><p><code><a href='https://rdrr.io/pkg/BSgenome/man/BSgenome-class.html'>BSgenome-class</a></code></p></td>
    </tr>
    <tr>
      <th>edits</th>
      <td><p>character vector: desired edits on '+' strand.
If named, names should be identical to those of <code>gr</code></p></td>
    </tr>
    <tr>
      <th>nprimer</th>
      <td><p>n primer nucleotides (default 13, max 17)</p></td>
    </tr>
    <tr>
      <th>nrt</th>
      <td><p>n rev transcr nucleotides (default 16, recomm. 10-16)</p></td>
    </tr>
    <tr>
      <th>ontargetmethod</th>
      <td><p>'Doench2014' or 'Doench2016': on-target scoring method</p></td>
    </tr>
    <tr>
      <th>offtargetmethod</th>
      <td><p>'bowtie' or 'pdict'</p></td>
    </tr>
    <tr>
      <th>mismatches</th>
      <td><p>no of primespacer mismatches 
(default 0, to suppress offtarget analysis: -1)</p></td>
    </tr>
    <tr>
      <th>nickmatches</th>
      <td><p>no of nickspacer offtarget mismatches 
(default 2, to suppresses offtarget analysis: -1)</p></td>
    </tr>
    <tr>
      <th>indexedgenomesdir</th>
      <td><p>directory with indexed genomes 
(as created by <code><a href='index_genome.html'>index_genome</a></code>)</p></td>
    </tr>
    <tr>
      <th>outdir</th>
      <td><p>directory whre offtarget analysis output is written</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>TRUE (default) or FALSE</p></td>
    </tr>
    <tr>
      <th>plot</th>
      <td><p>TRUE (default) or FALSE</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>passed to plot_intervals</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges-class</a></code> with prime editing spacer
ranges and following mcols: 
  * crisprspacer: N20 spacers
  * crisprpam:    NGG PAMs
  * crisprprimer: primer (on PAM strand)
  * crisprtranscript: reverse transcript (on PAM strand)
  * crisprextension:  3' extension of gRNA
              contains: reverse transcription template + primer binding site
              sequence can be found on non-PAM strand
  * crisprextrange: genomic range of crispr extension
  * Doench2016|4: on-target efficiency scores
  * off0, off1, off2: number of offtargets with 0, 1, 2 mismatches
  * off: total number of offtargets: off = off0 + off1 + ...
  * nickrange:        nickspacer range
  * nickspacer:       nickspacer sequence
  * nickDoench2016|4: nickspacer Doench scores
  * nickoff:          nickspacer offtarget counts</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Below the architecture of a prime editing site.
Edits can be performed anywhere in the revtranscript area.</p>
<p>spacer        pam
  --------------------===
          primer     revtranscript
      -------------================
  1..............17....GG..........
  .....................CC..........
      ----------extension----------</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='find_spacers.html'>find_spacers</a></code> to find standard crispr sites</p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Find PE spacers for 4 clinically relevant loci (Anzalone et al, 2019)</span>
    <span class='no'>bsgenome</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>BSgenome.Hsapiens.UCSC.hg38</span><span class='kw ns'>::</span><span class='no'><a href='https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg38/man/package.html'>BSgenome.Hsapiens.UCSC.hg38</a></span>
    <span class='no'>gr</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='char_to_granges.html'>char_to_granges</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(
        <span class='kw'>PRNP</span> <span class='kw'>=</span> <span class='st'>'chr20:4699600:+'</span>,             <span class='co'># snp: prion disease</span>
        <span class='kw'>HBB</span>  <span class='kw'>=</span> <span class='st'>'chr11:5227002:-'</span>,             <span class='co'># snp: sickle cell anemia</span>
        <span class='kw'>HEXA</span> <span class='kw'>=</span> <span class='st'>'chr15:72346580-72346583:-'</span>,   <span class='co'># del: tay sachs disease</span>
        <span class='kw'>CFTR</span> <span class='kw'>=</span> <span class='st'>'chr7:117559593-117559595:+'</span>), <span class='co'># ins: cystic fibrosis</span>
        <span class='no'>bsgenome</span>)
    <span class='no'>spacers</span> <span class='kw'>&lt;-</span> <span class='fu'>find_primespacers</span>(<span class='no'>gr</span>, <span class='no'>bsgenome</span>)</div><div class='output co'>#&gt; <span class='message'>Find primespacers for 4 targets</span></div><div class='output co'>#&gt; <span class='message'>	Found 10 primespacers</span></div><div class='output co'>#&gt; <span class='message'>	Count offtargets with max 0 mismatches using bowtie</span></div><div class='output co'>#&gt; <span class='message'>		Count genome (mis)matches</span></div><div class='output co'>#&gt; <span class='message'>			Expand iupac ambiguities in pam</span></div><div class='output co'>#&gt; <span class='message'>			Write reads to ~/multicrisprout/spacers.fa</span></div><div class='output co'>#&gt; <span class='message'>			Map reads: ~/multicrisprout/spacers/spacers_to_BSgenome.Hsapiens.UCSC.hg38.txt</span></div><div class='output co'>#&gt; <span class='message'>			Load results</span></div><div class='output co'>#&gt; <span class='message'>				Read 53 hits with max 1 mismatch(es)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 20 after removing NGG 'N' mismatches (avoid double counting expanded pams)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 20 after removing NGG 'G1' mismatches(except for NGG -&gt; NAG, which is allowed)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 20 after removing NGG 'G2' mismatches</span></div><div class='output co'>#&gt; <span class='message'>				Retain 11 after removing more than 0 mismatches</span></div><div class='output co'>#&gt; <span class='message'>			Count matches</span></div><div class='output co'>#&gt; <span class='message'>		Count off-targets</span></div><div class='output co'>#&gt; <span class='message'>		       9 are off0 free</span></div><div class='output co'>#&gt; <span class='message'>	Score ontargets</span></div><div class='output co'>#&gt; <span class='message'>		Add (4-23-3) contextseqs</span></div><div class='output co'>#&gt; <span class='message'>		Score contextseqs with Doench2014</span></div><div class='output co'>#&gt; <span class='message'>Find nickspacers</span></div><div class='output co'>#&gt; <span class='message'>Find spacers in 10 targets</span></div><div class='output co'>#&gt; <span class='message'>	Found 32 spacers</span></div><div class='output co'>#&gt; <span class='message'>	Count offtargets with max 2 mismatches using bowtie</span></div><div class='output co'>#&gt; <span class='message'>		Count genome (mis)matches</span></div><div class='output co'>#&gt; <span class='message'>			Expand iupac ambiguities in pam</span></div><div class='output co'>#&gt; <span class='message'>			Write reads to ~/multicrisprout/spacers.fa</span></div><div class='output co'>#&gt; <span class='message'>			Map reads: ~/multicrisprout/spacers/spacers_to_BSgenome.Hsapiens.UCSC.hg38.txt</span></div><div class='output co'>#&gt; <span class='message'>			Load results</span></div><div class='output co'>#&gt; <span class='message'>				Read 91 hits with max 2 mismatch(es)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 37 after removing NGG 'N' mismatches (avoid double counting expanded pams)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 36 after removing NGG 'G1' mismatches(except for NGG -&gt; NAG, which is allowed)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 33 after removing NGG 'G2' mismatches</span></div><div class='output co'>#&gt; <span class='message'>				Retain 33 after removing more than 2 mismatches</span></div><div class='output co'>#&gt; <span class='message'>			Count matches</span></div><div class='output co'>#&gt; <span class='message'>		Count off-targets</span></div><div class='output co'>#&gt; <span class='message'>		       30 are off0 free</span></div><div class='output co'>#&gt; <span class='message'>		       32 are off1 free</span></div><div class='output co'>#&gt; <span class='message'>		       24 are off2 free</span></div><div class='output co'>#&gt; <span class='message'>	Filter for best offtarget counts</span></div><div class='output co'>#&gt; <span class='message'>		Retain 32 with best off0 counts per pename</span></div><div class='output co'>#&gt; <span class='message'>		Retain 32 with best off1 counts per pename</span></div><div class='output co'>#&gt; <span class='message'>		Retain 24 with best off2 counts per pename</span></div><div class='output co'>#&gt; <span class='message'>	Score ontargets</span></div><div class='output co'>#&gt; <span class='message'>		Add (4-23-3) contextseqs</span></div><div class='output co'>#&gt; <span class='message'>		Score contextseqs with Doench2014</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Removed 34 rows containing missing values (geom_point).</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Removed 34 rows containing missing values (geom_point).</span></div><div class='img'><img src='find_primespacers-1.png' alt='' width='700' height='433' /></div><div class='input'>    <span class='no'>spacers</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='find_spacers.html'>find_spacers</a></span>(<span class='fu'><a href='extend_for_pe.html'>extend_for_pe</a></span>(<span class='no'>gr</span>), <span class='no'>bsgenome</span>, <span class='kw'>complement</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</div><div class='output co'>#&gt; <span class='message'>Find spacers in 8 targets</span></div><div class='output co'>#&gt; <span class='message'>	Found 10 spacers</span></div><div class='output co'>#&gt; <span class='message'>	Count offtargets with max 2 mismatches using bowtie</span></div><div class='output co'>#&gt; <span class='message'>		Count genome (mis)matches</span></div><div class='output co'>#&gt; <span class='message'>			Expand iupac ambiguities in pam</span></div><div class='output co'>#&gt; <span class='message'>			Write reads to ~/multicrisprout/spacers.fa</span></div><div class='output co'>#&gt; <span class='message'>			Map reads: ~/multicrisprout/spacers/spacers_to_BSgenome.Hsapiens.UCSC.hg38.txt</span></div><div class='output co'>#&gt; <span class='message'>			Load results</span></div><div class='output co'>#&gt; <span class='message'>				Read 140 hits with max 2 mismatch(es)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 80 after removing NGG 'N' mismatches (avoid double counting expanded pams)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 77 after removing NGG 'G1' mismatches(except for NGG -&gt; NAG, which is allowed)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 71 after removing NGG 'G2' mismatches</span></div><div class='output co'>#&gt; <span class='message'>				Retain 71 after removing more than 2 mismatches</span></div><div class='output co'>#&gt; <span class='message'>			Count matches</span></div><div class='output co'>#&gt; <span class='message'>		Count off-targets</span></div><div class='output co'>#&gt; <span class='message'>		       9 are off0 free</span></div><div class='output co'>#&gt; <span class='message'>		       9 are off1 free</span></div><div class='output co'>#&gt; <span class='message'>		       2 are off2 free</span></div><div class='output co'>#&gt; <span class='message'>	Score ontargets</span></div><div class='output co'>#&gt; <span class='message'>		Add (4-23-3) contextseqs</span></div><div class='output co'>#&gt; <span class='message'>		Score contextseqs with Doench2014</span></div><div class='img'><img src='find_primespacers-2.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># Edit PRNP locus for resistance against prion disease (Anzalone et al, 2019)</span>
    <span class='no'>bsgenome</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>BSgenome.Hsapiens.UCSC.hg38</span><span class='kw ns'>::</span><span class='no'><a href='https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg38/man/package.html'>BSgenome.Hsapiens.UCSC.hg38</a></span>
    <span class='no'>gr</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='char_to_granges.html'>char_to_granges</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='kw'>PRNP</span> <span class='kw'>=</span> <span class='st'>'chr20:4699600:+'</span>), <span class='no'>bsgenome</span>)
    <span class='fu'>find_primespacers</span>(<span class='no'>gr</span>, <span class='no'>bsgenome</span>)</div><div class='output co'>#&gt; <span class='message'>Find primespacers for 1 targets</span></div><div class='output co'>#&gt; <span class='message'>	Found 4 primespacers</span></div><div class='output co'>#&gt; <span class='message'>	Count offtargets with max 0 mismatches using bowtie</span></div><div class='output co'>#&gt; <span class='message'>		Count genome (mis)matches</span></div><div class='output co'>#&gt; <span class='message'>			Expand iupac ambiguities in pam</span></div><div class='output co'>#&gt; <span class='message'>			Write reads to ~/multicrisprout/spacers.fa</span></div><div class='output co'>#&gt; <span class='message'>			Map reads: ~/multicrisprout/spacers/spacers_to_BSgenome.Hsapiens.UCSC.hg38.txt</span></div><div class='output co'>#&gt; <span class='message'>			Load results</span></div><div class='output co'>#&gt; <span class='message'>				Read 25 hits with max 1 mismatch(es)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 13 after removing NGG 'N' mismatches (avoid double counting expanded pams)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 13 after removing NGG 'G1' mismatches(except for NGG -&gt; NAG, which is allowed)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 13 after removing NGG 'G2' mismatches</span></div><div class='output co'>#&gt; <span class='message'>				Retain 4 after removing more than 0 mismatches</span></div><div class='output co'>#&gt; <span class='message'>			Count matches</span></div><div class='output co'>#&gt; <span class='message'>		Count off-targets</span></div><div class='output co'>#&gt; <span class='message'>		       4 are off0 free</span></div><div class='output co'>#&gt; <span class='message'>	Score ontargets</span></div><div class='output co'>#&gt; <span class='message'>		Add (4-23-3) contextseqs</span></div><div class='output co'>#&gt; <span class='message'>		Score contextseqs with Doench2014</span></div><div class='output co'>#&gt; <span class='message'>Find nickspacers</span></div><div class='output co'>#&gt; <span class='message'>Find spacers in 4 targets</span></div><div class='output co'>#&gt; <span class='message'>	Found 12 spacers</span></div><div class='output co'>#&gt; <span class='message'>	Count offtargets with max 2 mismatches using bowtie</span></div><div class='output co'>#&gt; <span class='message'>		Count genome (mis)matches</span></div><div class='output co'>#&gt; <span class='message'>			Expand iupac ambiguities in pam</span></div><div class='output co'>#&gt; <span class='message'>			Write reads to ~/multicrisprout/spacers.fa</span></div><div class='output co'>#&gt; <span class='message'>			Map reads: ~/multicrisprout/spacers/spacers_to_BSgenome.Hsapiens.UCSC.hg38.txt</span></div><div class='output co'>#&gt; <span class='message'>			Load results</span></div><div class='output co'>#&gt; <span class='message'>				Read 13 hits with max 2 mismatch(es)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 4 after removing NGG 'N' mismatches (avoid double counting expanded pams)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 4 after removing NGG 'G1' mismatches(except for NGG -&gt; NAG, which is allowed)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 3 after removing NGG 'G2' mismatches</span></div><div class='output co'>#&gt; <span class='message'>				Retain 3 after removing more than 2 mismatches</span></div><div class='output co'>#&gt; <span class='message'>			Count matches</span></div><div class='output co'>#&gt; <span class='message'>		Count off-targets</span></div><div class='output co'>#&gt; <span class='message'>		       12 are off0 free</span></div><div class='output co'>#&gt; <span class='message'>		       12 are off1 free</span></div><div class='output co'>#&gt; <span class='message'>		       12 are off2 free</span></div><div class='output co'>#&gt; <span class='message'>	Filter for best offtarget counts</span></div><div class='output co'>#&gt; <span class='message'>		Retain 12 with best off0 counts per pename</span></div><div class='output co'>#&gt; <span class='message'>		Retain 12 with best off1 counts per pename</span></div><div class='output co'>#&gt; <span class='message'>		Retain 12 with best off2 counts per pename</span></div><div class='output co'>#&gt; <span class='message'>	Score ontargets</span></div><div class='output co'>#&gt; <span class='message'>		Add (4-23-3) contextseqs</span></div><div class='output co'>#&gt; <span class='message'>		Score contextseqs with Doench2014</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Removed 16 rows containing missing values (geom_point).</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Removed 16 rows containing missing values (geom_point).</span></div><div class='img'><img src='find_primespacers-3.png' alt='' width='700' height='433' /></div><div class='output co'>#&gt; GRanges object with 4 ranges and 21 metadata columns:
#&gt;          seqnames          ranges strand |  targetname targetstart targetend
#&gt;             &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;   &lt;integer&gt; &lt;integer&gt;
#&gt;   PRNP_1    chr20 4699568-4699587      + |        PRNP     4699600   4699600
#&gt;   PRNP_2    chr20 4699569-4699588      + |        PRNP     4699600   4699600
#&gt;   PRNP_3    chr20 4699575-4699594      + |        PRNP     4699600   4699600
#&gt;   PRNP_4    chr20 4699578-4699597      + |        PRNP     4699600   4699600
#&gt;           crisprname         crisprspacer   crisprpam  crisprprimer
#&gt;          &lt;character&gt;          &lt;character&gt; &lt;character&gt;   &lt;character&gt;
#&gt;   PRNP_1      PRNP_1 AGCAGCTGGGGCAGTGGTGG         GGG GCTGGGGCAGTGG
#&gt;   PRNP_2      PRNP_2 GCAGCTGGGGCAGTGGTGGG         GGG CTGGGGCAGTGGT
#&gt;   PRNP_3      PRNP_3 GGGGCAGTGGTGGGGGGCCT         TGG CAGTGGTGGGGGG
#&gt;   PRNP_4      PRNP_4 GCAGTGGTGGGGGGCCTTGG         CGG TGGTGGGGGGCCT
#&gt;          crisprtranscript               crisprextension          crisprextrange
#&gt;               &lt;character&gt;                   &lt;character&gt;             &lt;character&gt;
#&gt;   PRNP_1 TGGGGGGCCTTGGCGG CCGCCAAGGCCCCCCACCACTGCCCCAGC chr20:4699572-4699600:-
#&gt;   PRNP_2 GGGGGGCCTTGGCGGC GCCGCCAAGGCCCCCCACCACTGCCCCAG chr20:4699573-4699601:-
#&gt;   PRNP_3 CCTTGGCGGCTACATG CATGTAGCCGCCAAGGCCCCCCACCACTG chr20:4699579-4699607:-
#&gt;   PRNP_4 TGGCGGCTACATGCTG CAGCATGTAGCCGCCAAGGCCCCCCACCA chr20:4699582-4699610:-
#&gt;                off      off0 Doench2014
#&gt;          &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;
#&gt;   PRNP_1         0         0 0.00994899
#&gt;   PRNP_2         0         0 0.04496627
#&gt;   PRNP_3         0         0 0.00996254
#&gt;   PRNP_4         0         0 0.20935863
#&gt;                                                                        nickrange
#&gt;                                                                      &lt;character&gt;
#&gt;   PRNP_1 chr20:4699632-4699651:-;chr20:4699633-4699652:-;chr20:4699664-4699683:-
#&gt;   PRNP_2 chr20:4699632-4699651:-;chr20:4699633-4699652:-;chr20:4699664-4699683:-
#&gt;   PRNP_3 chr20:4699632-4699651:-;chr20:4699633-4699652:-;chr20:4699664-4699683:-
#&gt;   PRNP_4 chr20:4699632-4699651:-;chr20:4699633-4699652:-;chr20:4699664-4699683:-
#&gt;                                                              nickspacer
#&gt;                                                             &lt;character&gt;
#&gt;   PRNP_1 TCACTGCCGAAATGTATGAT;GTCACTGCCGAAATGTATGA;GCATGTTTTCACGATAGTAA
#&gt;   PRNP_2 TCACTGCCGAAATGTATGAT;GTCACTGCCGAAATGTATGA;GCATGTTTTCACGATAGTAA
#&gt;   PRNP_3 TCACTGCCGAAATGTATGAT;GTCACTGCCGAAATGTATGA;GCATGTTTTCACGATAGTAA
#&gt;   PRNP_4 TCACTGCCGAAATGTATGAT;GTCACTGCCGAAATGTATGA;GCATGTTTTCACGATAGTAA
#&gt;              nickpam     nickoff    nickoff0    nickoff1    nickoff2
#&gt;          &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;
#&gt;   PRNP_1 GGG;TGG;CGG       0;0;0       0;0;0       0;0;0       0;0;0
#&gt;   PRNP_2 GGG;TGG;CGG       0;0;0       0;0;0       0;0;0       0;0;0
#&gt;   PRNP_3 GGG;TGG;CGG       0;0;0       0;0;0       0;0;0       0;0;0
#&gt;   PRNP_4 GGG;TGG;CGG       0;0;0       0;0;0       0;0;0       0;0;0
#&gt;          nickDoench2014
#&gt;             &lt;character&gt;
#&gt;   PRNP_1 0.14;0.32;0.52
#&gt;   PRNP_2 0.14;0.32;0.52
#&gt;   PRNP_3 0.14;0.32;0.52
#&gt;   PRNP_4 0.14;0.32;0.52
#&gt;   -------
#&gt;   seqinfo: 595 sequences (1 circular) from hg38 genome</div><div class='input'>    <span class='fu'>find_primespacers</span>(<span class='no'>gr</span>, <span class='no'>bsgenome</span>, <span class='kw'>edits</span> <span class='kw'>=</span> <span class='st'>'T'</span>)</div><div class='output co'>#&gt; <span class='message'>Find primespacers for 1 targets</span></div><div class='output co'>#&gt; <span class='message'>	Found 4 primespacers</span></div><div class='output co'>#&gt; <span class='message'>	Count offtargets with max 0 mismatches using bowtie</span></div><div class='output co'>#&gt; <span class='message'>		Count genome (mis)matches</span></div><div class='output co'>#&gt; <span class='message'>			Expand iupac ambiguities in pam</span></div><div class='output co'>#&gt; <span class='message'>			Write reads to ~/multicrisprout/spacers.fa</span></div><div class='output co'>#&gt; <span class='message'>			Map reads: ~/multicrisprout/spacers/spacers_to_BSgenome.Hsapiens.UCSC.hg38.txt</span></div><div class='output co'>#&gt; <span class='message'>			Load results</span></div><div class='output co'>#&gt; <span class='message'>				Read 25 hits with max 1 mismatch(es)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 13 after removing NGG 'N' mismatches (avoid double counting expanded pams)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 13 after removing NGG 'G1' mismatches(except for NGG -&gt; NAG, which is allowed)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 13 after removing NGG 'G2' mismatches</span></div><div class='output co'>#&gt; <span class='message'>				Retain 4 after removing more than 0 mismatches</span></div><div class='output co'>#&gt; <span class='message'>			Count matches</span></div><div class='output co'>#&gt; <span class='message'>		Count off-targets</span></div><div class='output co'>#&gt; <span class='message'>		       4 are off0 free</span></div><div class='output co'>#&gt; <span class='message'>	Score ontargets</span></div><div class='output co'>#&gt; <span class='message'>		Add (4-23-3) contextseqs</span></div><div class='output co'>#&gt; <span class='message'>		Score contextseqs with Doench2014</span></div><div class='output co'>#&gt; <span class='message'>Find nickspacers</span></div><div class='output co'>#&gt; <span class='message'>Find spacers in 4 targets</span></div><div class='output co'>#&gt; <span class='message'>	Found 12 spacers</span></div><div class='output co'>#&gt; <span class='message'>	Count offtargets with max 2 mismatches using bowtie</span></div><div class='output co'>#&gt; <span class='message'>		Count genome (mis)matches</span></div><div class='output co'>#&gt; <span class='message'>			Expand iupac ambiguities in pam</span></div><div class='output co'>#&gt; <span class='message'>			Write reads to ~/multicrisprout/spacers.fa</span></div><div class='output co'>#&gt; <span class='message'>			Map reads: ~/multicrisprout/spacers/spacers_to_BSgenome.Hsapiens.UCSC.hg38.txt</span></div><div class='output co'>#&gt; <span class='message'>			Load results</span></div><div class='output co'>#&gt; <span class='message'>				Read 13 hits with max 2 mismatch(es)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 4 after removing NGG 'N' mismatches (avoid double counting expanded pams)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 4 after removing NGG 'G1' mismatches(except for NGG -&gt; NAG, which is allowed)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 3 after removing NGG 'G2' mismatches</span></div><div class='output co'>#&gt; <span class='message'>				Retain 3 after removing more than 2 mismatches</span></div><div class='output co'>#&gt; <span class='message'>			Count matches</span></div><div class='output co'>#&gt; <span class='message'>		Count off-targets</span></div><div class='output co'>#&gt; <span class='message'>		       12 are off0 free</span></div><div class='output co'>#&gt; <span class='message'>		       12 are off1 free</span></div><div class='output co'>#&gt; <span class='message'>		       12 are off2 free</span></div><div class='output co'>#&gt; <span class='message'>	Filter for best offtarget counts</span></div><div class='output co'>#&gt; <span class='message'>		Retain 12 with best off0 counts per pename</span></div><div class='output co'>#&gt; <span class='message'>		Retain 12 with best off1 counts per pename</span></div><div class='output co'>#&gt; <span class='message'>		Retain 12 with best off2 counts per pename</span></div><div class='output co'>#&gt; <span class='message'>	Score ontargets</span></div><div class='output co'>#&gt; <span class='message'>		Add (4-23-3) contextseqs</span></div><div class='output co'>#&gt; <span class='message'>		Score contextseqs with Doench2014</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Removed 16 rows containing missing values (geom_point).</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Removed 16 rows containing missing values (geom_point).</span></div><div class='img'><img src='find_primespacers-4.png' alt='' width='700' height='433' /></div><div class='output co'>#&gt; GRanges object with 4 ranges and 21 metadata columns:
#&gt;          seqnames          ranges strand |  targetname targetstart targetend
#&gt;             &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;   &lt;integer&gt; &lt;integer&gt;
#&gt;   PRNP_1    chr20 4699568-4699587      + |        PRNP     4699600   4699600
#&gt;   PRNP_2    chr20 4699569-4699588      + |        PRNP     4699600   4699600
#&gt;   PRNP_3    chr20 4699575-4699594      + |        PRNP     4699600   4699600
#&gt;   PRNP_4    chr20 4699578-4699597      + |        PRNP     4699600   4699600
#&gt;           crisprname         crisprspacer   crisprpam  crisprprimer
#&gt;          &lt;character&gt;          &lt;character&gt; &lt;character&gt;   &lt;character&gt;
#&gt;   PRNP_1      PRNP_1 AGCAGCTGGGGCAGTGGTGG         GGG GCTGGGGCAGTGG
#&gt;   PRNP_2      PRNP_2 GCAGCTGGGGCAGTGGTGGG         GGG CTGGGGCAGTGGT
#&gt;   PRNP_3      PRNP_3 GGGGCAGTGGTGGGGGGCCT         TGG CAGTGGTGGGGGG
#&gt;   PRNP_4      PRNP_4 GCAGTGGTGGGGGGCCTTGG         CGG TGGTGGGGGGCCT
#&gt;          crisprtranscript               crisprextension          crisprextrange
#&gt;               &lt;character&gt;                   &lt;character&gt;             &lt;character&gt;
#&gt;   PRNP_1 TGGGGGGCCTTGGCGT ACGCCAAGGCCCCCCACCACTGCCCCAGC chr20:4699572-4699600:-
#&gt;   PRNP_2 GGGGGGCCTTGGCGTC GACGCCAAGGCCCCCCACCACTGCCCCAG chr20:4699573-4699601:-
#&gt;   PRNP_3 CCTTGGCGTCTACATG CATGTAGACGCCAAGGCCCCCCACCACTG chr20:4699579-4699607:-
#&gt;   PRNP_4 TGGCGTCTACATGCTG CAGCATGTAGACGCCAAGGCCCCCCACCA chr20:4699582-4699610:-
#&gt;                off      off0 Doench2014
#&gt;          &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;
#&gt;   PRNP_1         0         0 0.00994899
#&gt;   PRNP_2         0         0 0.04496627
#&gt;   PRNP_3         0         0 0.00996254
#&gt;   PRNP_4         0         0 0.20935863
#&gt;                                                                        nickrange
#&gt;                                                                      &lt;character&gt;
#&gt;   PRNP_1 chr20:4699632-4699651:-;chr20:4699633-4699652:-;chr20:4699664-4699683:-
#&gt;   PRNP_2 chr20:4699632-4699651:-;chr20:4699633-4699652:-;chr20:4699664-4699683:-
#&gt;   PRNP_3 chr20:4699632-4699651:-;chr20:4699633-4699652:-;chr20:4699664-4699683:-
#&gt;   PRNP_4 chr20:4699632-4699651:-;chr20:4699633-4699652:-;chr20:4699664-4699683:-
#&gt;                                                              nickspacer
#&gt;                                                             &lt;character&gt;
#&gt;   PRNP_1 TCACTGCCGAAATGTATGAT;GTCACTGCCGAAATGTATGA;GCATGTTTTCACGATAGTAA
#&gt;   PRNP_2 TCACTGCCGAAATGTATGAT;GTCACTGCCGAAATGTATGA;GCATGTTTTCACGATAGTAA
#&gt;   PRNP_3 TCACTGCCGAAATGTATGAT;GTCACTGCCGAAATGTATGA;GCATGTTTTCACGATAGTAA
#&gt;   PRNP_4 TCACTGCCGAAATGTATGAT;GTCACTGCCGAAATGTATGA;GCATGTTTTCACGATAGTAA
#&gt;              nickpam     nickoff    nickoff0    nickoff1    nickoff2
#&gt;          &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;
#&gt;   PRNP_1 GGG;TGG;CGG       0;0;0       0;0;0       0;0;0       0;0;0
#&gt;   PRNP_2 GGG;TGG;CGG       0;0;0       0;0;0       0;0;0       0;0;0
#&gt;   PRNP_3 GGG;TGG;CGG       0;0;0       0;0;0       0;0;0       0;0;0
#&gt;   PRNP_4 GGG;TGG;CGG       0;0;0       0;0;0       0;0;0       0;0;0
#&gt;          nickDoench2014
#&gt;             &lt;character&gt;
#&gt;   PRNP_1 0.14;0.32;0.52
#&gt;   PRNP_2 0.14;0.32;0.52
#&gt;   PRNP_3 0.14;0.32;0.52
#&gt;   PRNP_4 0.14;0.32;0.52
#&gt;   -------
#&gt;   seqinfo: 595 sequences (1 circular) from hg38 genome</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Aditya Bhagwat.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


