<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Convert geneids into GRanges — genes_to_granges • multicrispr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Convert geneids into GRanges — genes_to_granges" />
<meta property="og:description" content="Convert geneids into GRanges" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">multicrispr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.24</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/crispr_grna_design.html">multicrispr: Crispr/Cas9 gRNA design</a>
    </li>
    <li>
      <a href="../articles/genome_arithmetics.html">Genome arithmetics with multicrispr</a>
    </li>
    <li>
      <a href="../articles/prime_editing.html">multicrispr: Prime Editing</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Convert geneids into GRanges</h1>
    
    <div class="hidden name"><code>genes_to_granges.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Convert geneids into GRanges</p>
    </div>

    <pre class="usage"><span class='fu'>genes_to_granges</span>(<span class='no'>geneids</span>, <span class='no'>txdb</span>, <span class='kw'>complement</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>plot</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='fu'>genefile_to_granges</span>(<span class='no'>file</span>, <span class='no'>txdb</span>, <span class='kw'>complement</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>plot</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>geneids</th>
      <td><p>Gene identifier vector</p></td>
    </tr>
    <tr>
      <th>txdb</th>
      <td><p><code><a href='https://rdrr.io/pkg/GenomicFeatures/man/TxDb-class.html'>TxDb-class</a></code> or 
<code><a href='https://rdrr.io/pkg/ensembldb/man/EnsDb-class.html'>EnsDb-class</a></code></p></td>
    </tr>
    <tr>
      <th>complement</th>
      <td><p>TRUE (default) or FALSE: add complementary strand?</p></td>
    </tr>
    <tr>
      <th>plot</th>
      <td><p>TRUE (default) or FALSE</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>TRUE (default) or FALSE</p></td>
    </tr>
    <tr>
      <th>file</th>
      <td><p>Gene identifier file (one per row)</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges-class</a></code></p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='char_to_granges.html'>char_to_granges</a></code>, <code><a href='bed_to_granges.html'>bed_to_granges</a></code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Entrez</span>
<span class='co'>#-------</span>
    <span class='no'>genefile</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'extdata/SRF.entrez'</span>, <span class='kw'>package</span><span class='kw'>=</span><span class='st'>'multicrispr'</span>)
    <span class='no'>geneids</span>  <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/character.html'>as.character</a></span>(<span class='fu'><a href='https://rdrr.io/r/utils/read.table.html'>read.table</a></span>(<span class='no'>genefile</span>)<span class='kw'>[[</span><span class='fl'>1</span>]])
    <span class='no'>txdb</span>     <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/utils/getFromNamespace.html'>getFromNamespace</a></span>(<span class='st'>'TxDb.Mmusculus.UCSC.mm10.knownGene'</span>,
                             <span class='st'>'TxDb.Mmusculus.UCSC.mm10.knownGene'</span>)
    (<span class='no'>gr</span> <span class='kw'>&lt;-</span> <span class='fu'>genes_to_granges</span>(<span class='no'>geneids</span>, <span class='no'>txdb</span>))</div><div class='output co'>#&gt; <span class='message'>  66 genes were dropped because they have exons located on both strands</span>
#&gt; <span class='message'>  of the same reference sequence or on more than one reference sequence,</span>
#&gt; <span class='message'>  so cannot be represented by a single genomic range.</span>
#&gt; <span class='message'>  Use 'single.strand.genes.only=FALSE' to get all the genes in a</span>
#&gt; <span class='message'>  GRangesList object, or use suppressMessages() to suppress this message.</span></div><div class='output co'>#&gt; <span class='message'>		Convert 10 genes to 10 GRanges</span></div><div class='output co'>#&gt; <span class='message'>	Retain 20 unique target ranges after adding inverse strands</span></div><div class='output co'>#&gt; GRanges object with 20 ranges and 4 metadata columns:
#&gt;               seqnames              ranges strand |     gene_id  targetname
#&gt;                  &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;
#&gt;       27392_r     chr1 105518422-105663677      - |       27392     27392_r
#&gt;       27392_f     chr1 105518422-105663677      + |       27392     27392_f
#&gt;   100043424_r     chr2 128178319-128502765      - |   100043424 100043424_r
#&gt;   100043424_f     chr2 128178319-128502765      + |   100043424 100043424_f
#&gt;      622434_f     chr3   62338344-62462221      + |      622434    622434_f
#&gt;           ...      ...                 ...    ... .         ...         ...
#&gt;       71918_f    chr14   25711642-25769039      + |       71918     71918_f
#&gt;       13653_f    chr18   34859823-34864984      + |       13653     13653_f
#&gt;       13653_r    chr18   34859823-34864984      - |       13653     13653_r
#&gt;       52662_f    chr18   67933257-68268630      + |       52662     52662_f
#&gt;       52662_r    chr18   67933257-68268630      - |       52662     52662_r
#&gt;               targetstart targetend
#&gt;                 &lt;integer&gt; &lt;integer&gt;
#&gt;       27392_r   105518422 105663677
#&gt;       27392_f   105518422 105663677
#&gt;   100043424_r   128178319 128502765
#&gt;   100043424_f   128178319 128502765
#&gt;      622434_f    62338344  62462221
#&gt;           ...         ...       ...
#&gt;       71918_f    25711642  25769039
#&gt;       13653_f    34859823  34864984
#&gt;       13653_r    34859823  34864984
#&gt;       52662_f    67933257  68268630
#&gt;       52662_r    67933257  68268630
#&gt;   -------
#&gt;   seqinfo: 66 sequences (1 circular) from mm10 genome</div><div class='input'>    (<span class='no'>gr</span> <span class='kw'>&lt;-</span> <span class='fu'>genefile_to_granges</span>(<span class='no'>genefile</span>, <span class='no'>txdb</span>))</div><div class='output co'>#&gt; <span class='message'>  66 genes were dropped because they have exons located on both strands</span>
#&gt; <span class='message'>  of the same reference sequence or on more than one reference sequence,</span>
#&gt; <span class='message'>  so cannot be represented by a single genomic range.</span>
#&gt; <span class='message'>  Use 'single.strand.genes.only=FALSE' to get all the genes in a</span>
#&gt; <span class='message'>  GRangesList object, or use suppressMessages() to suppress this message.</span></div><div class='output co'>#&gt; <span class='message'>		Convert 10 genes to 10 GRanges</span></div><div class='output co'>#&gt; <span class='message'>	Retain 20 unique target ranges after adding inverse strands</span></div><div class='img'><img src='genes_to_granges-1.png' alt='' width='700' height='433' /></div><div class='output co'>#&gt; GRanges object with 20 ranges and 4 metadata columns:
#&gt;               seqnames              ranges strand |     gene_id  targetname
#&gt;                  &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;
#&gt;       27392_r     chr1 105518422-105663677      - |       27392     27392_r
#&gt;       27392_f     chr1 105518422-105663677      + |       27392     27392_f
#&gt;   100043424_r     chr2 128178319-128502765      - |   100043424 100043424_r
#&gt;   100043424_f     chr2 128178319-128502765      + |   100043424 100043424_f
#&gt;      622434_f     chr3   62338344-62462221      + |      622434    622434_f
#&gt;           ...      ...                 ...    ... .         ...         ...
#&gt;       71918_f    chr14   25711642-25769039      + |       71918     71918_f
#&gt;       13653_f    chr18   34859823-34864984      + |       13653     13653_f
#&gt;       13653_r    chr18   34859823-34864984      - |       13653     13653_r
#&gt;       52662_f    chr18   67933257-68268630      + |       52662     52662_f
#&gt;       52662_r    chr18   67933257-68268630      - |       52662     52662_r
#&gt;               targetstart targetend
#&gt;                 &lt;integer&gt; &lt;integer&gt;
#&gt;       27392_r   105518422 105663677
#&gt;       27392_f   105518422 105663677
#&gt;   100043424_r   128178319 128502765
#&gt;   100043424_f   128178319 128502765
#&gt;      622434_f    62338344  62462221
#&gt;           ...         ...       ...
#&gt;       71918_f    25711642  25769039
#&gt;       13653_f    34859823  34864984
#&gt;       13653_r    34859823  34864984
#&gt;       52662_f    67933257  68268630
#&gt;       52662_r    67933257  68268630
#&gt;   -------
#&gt;   seqinfo: 66 sequences (1 circular) from mm10 genome</div><div class='input'>
# Ensembl
#--------
    # txdb &lt;- AnnotationHub::AnnotationHub()[["AH75036"]]
    # genefile &lt;- system.file('extdata/SRF.ensembl', package='multicrispr')
    # geneids &lt;- as.character(read.table(genefile)[[1]])
    # (gr &lt;- genes_to_granges(geneids, txdb))
    # (gr &lt;- genefile_to_granges(genefile, txdb))
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Aditya Bhagwat.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


