<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Interval plot GRanges — plot_intervals • multicrispr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Interval plot GRanges — plot_intervals" />
<meta property="og:description" content="Interval plot GRanges" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">multicrispr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.23</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/genome_arithmetics.html">Genome arithmetics with multicrispr</a>
    </li>
    <li>
      <a href="../articles/grna_design.html">Using multicrispr for gRNA design</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Interval plot GRanges</h1>
    
    <div class="hidden name"><code>plot_intervals.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Interval plot GRanges</p>
    </div>

    <pre class="usage"><span class='fu'>plot_intervals</span>(
  <span class='no'>gr</span>,
  <span class='kw'>xref</span> <span class='kw'>=</span> <span class='st'>"targetname"</span>,
  <span class='kw'>y</span> <span class='kw'>=</span> <span class='fu'>default_y</span>(<span class='no'>gr</span>),
  <span class='kw'>nperchrom</span> <span class='kw'>=</span> <span class='fl'>2</span>,
  <span class='kw'>nchrom</span> <span class='kw'>=</span> <span class='fl'>4</span>,
  <span class='kw'>color_var</span> <span class='kw'>=</span> <span class='st'>"targetname"</span>,
  <span class='kw'>facet_var</span> <span class='kw'>=</span> <span class='st'>"seqnames"</span>,
  <span class='kw'>linetype_var</span> <span class='kw'>=</span> <span class='fu'>default_linetype</span>(<span class='no'>gr</span>),
  <span class='kw'>size_var</span> <span class='kw'>=</span> <span class='fu'>default_size_var</span>(<span class='no'>gr</span>),
  <span class='kw'>alpha_var</span> <span class='kw'>=</span> <span class='fu'>default_alpha_var</span>(<span class='no'>gr</span>),
  <span class='kw'>title</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>scales</span> <span class='kw'>=</span> <span class='st'>"free"</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>gr</th>
      <td><p><code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges-class</a></code></p></td>
    </tr>
    <tr>
      <th>xref</th>
      <td><p>gr var used for scaling x axis</p></td>
    </tr>
    <tr>
      <th>y</th>
      <td><p>'names' (default) or name of gr variable</p></td>
    </tr>
    <tr>
      <th>nperchrom</th>
      <td><p>number (default 1): n head (and n tail) targets 
shown per chromosome</p></td>
    </tr>
    <tr>
      <th>nchrom</th>
      <td><p>number (default 6) of chromosomes shown</p></td>
    </tr>
    <tr>
      <th>color_var</th>
      <td><p>'seqnames' (default) or other gr variable</p></td>
    </tr>
    <tr>
      <th>facet_var</th>
      <td><p>NULL(default)  or gr variable mapped to facet</p></td>
    </tr>
    <tr>
      <th>linetype_var</th>
      <td><p>NULL (default) or gr variable mapped to linetype</p></td>
    </tr>
    <tr>
      <th>size_var</th>
      <td><p>NULL (default) or gr variable mapped to size</p></td>
    </tr>
    <tr>
      <th>alpha_var</th>
      <td><p>NULL or gr variable mapped to alpha</p></td>
    </tr>
    <tr>
      <th>title</th>
      <td><p>NULL or string: plot title</p></td>
    </tr>
    <tr>
      <th>scales</th>
      <td><p>'free', 'fixed', etc</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>ggplot object</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='plot_karyogram.html'>plot_karyogram</a></code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># SRF sites</span>
    <span class='fu'><a href='https://rdrr.io/r/base/library.html'>require</a></span>(<span class='no'>magrittr</span>)
    <span class='no'>bsgenome</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>BSgenome.Mmusculus.UCSC.mm10</span><span class='kw ns'>::</span><span class='no'><a href='https://rdrr.io/pkg/BSgenome.Mmusculus.UCSC.mm10/man/package.html'>BSgenome.Mmusculus.UCSC.mm10</a></span>
    <span class='no'>bedfile</span> <span class='kw'>&lt;-</span>  <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'extdata/SRF.bed'</span>,  <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'multicrispr'</span>)
    <span class='no'>targets</span>   <span class='kw'>&lt;-</span> <span class='fu'><a href='bed_to_granges.html'>bed_to_granges</a></span>(<span class='no'>bedfile</span>, <span class='st'>'mm10'</span>, <span class='kw'>plot</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</div><div class='output co'>#&gt; <span class='message'>	Read SRF.bed into GRanges</span></div><div class='output co'>#&gt; <span class='message'>		1974 ranges on 21 chromosomes</span></div><div class='input'>    <span class='fu'>plot_intervals</span>(<span class='no'>targets</span>)</div><div class='img'><img src='plot_intervals-1.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># PE targets</span>
    <span class='no'>bsgenome</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>BSgenome.Hsapiens.UCSC.hg38</span><span class='kw ns'>::</span><span class='no'><a href='https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg38/man/package.html'>BSgenome.Hsapiens.UCSC.hg38</a></span>
    <span class='no'>gr</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='char_to_granges.html'>char_to_granges</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='kw'>PRNP</span> <span class='kw'>=</span> <span class='st'>'chr20:4699600:+'</span>,
                            <span class='kw'>HBB</span>  <span class='kw'>=</span> <span class='st'>'chr11:5227002:-'</span>,
                            <span class='kw'>HEXA</span> <span class='kw'>=</span> <span class='st'>'chr15:72346580-72346583:-'</span>,
                            <span class='kw'>CFTR</span> <span class='kw'>=</span> <span class='st'>'chr7:117559593-117559595:+'</span>),
                          <span class='no'>bsgenome</span>)
    <span class='no'>spacers</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='find_primespacers.html'>find_primespacers</a></span>(<span class='no'>gr</span>, <span class='no'>bsgenome</span>, <span class='kw'>plot</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</div><div class='output co'>#&gt; <span class='message'>Find primespacers for 4 targets</span></div><div class='output co'>#&gt; <span class='message'>	Found 10 primespacers</span></div><div class='output co'>#&gt; <span class='message'>	Count offtargets with max 0 mismatches using bowtie</span></div><div class='output co'>#&gt; <span class='message'>		Count genome (mis)matches</span></div><div class='output co'>#&gt; <span class='message'>			Expand iupac ambiguities in pam</span></div><div class='output co'>#&gt; <span class='message'>			Write reads to ~/multicrisprout/spacers.fa</span></div><div class='output co'>#&gt; <span class='message'>			Map reads: ~/multicrisprout/spacers/spacers_to_BSgenome.Hsapiens.UCSC.hg38.txt</span></div><div class='output co'>#&gt; <span class='message'>			Load results</span></div><div class='output co'>#&gt; <span class='message'>				Read 53 hits with max 1 mismatch(es)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 20 after removing NGG 'N' mismatches (avoid double counting expanded pams)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 20 after removing NGG 'G1' mismatches(except for NGG -&gt; NAG, which is allowed)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 20 after removing NGG 'G2' mismatches</span></div><div class='output co'>#&gt; <span class='message'>				Retain 11 after removing more than 0 mismatches</span></div><div class='output co'>#&gt; <span class='message'>			Count matches</span></div><div class='output co'>#&gt; <span class='message'>		Count off-targets</span></div><div class='output co'>#&gt; <span class='message'>		       9 are off0 free</span></div><div class='output co'>#&gt; <span class='message'>	Score ontargets</span></div><div class='output co'>#&gt; <span class='message'>		Add (4-23-3) contextseqs</span></div><div class='output co'>#&gt; <span class='message'>		Score contextseqs with Doench2014</span></div><div class='output co'>#&gt; <span class='message'>Find nickspacers</span></div><div class='output co'>#&gt; <span class='message'>Find spacers in 10 targets</span></div><div class='output co'>#&gt; <span class='message'>	Found 32 spacers</span></div><div class='output co'>#&gt; <span class='message'>	Count offtargets with max 2 mismatches using bowtie</span></div><div class='output co'>#&gt; <span class='message'>		Count genome (mis)matches</span></div><div class='output co'>#&gt; <span class='message'>			Expand iupac ambiguities in pam</span></div><div class='output co'>#&gt; <span class='message'>			Write reads to ~/multicrisprout/spacers.fa</span></div><div class='output co'>#&gt; <span class='message'>			Map reads: ~/multicrisprout/spacers/spacers_to_BSgenome.Hsapiens.UCSC.hg38.txt</span></div><div class='output co'>#&gt; <span class='message'>			Load results</span></div><div class='output co'>#&gt; <span class='message'>				Read 91 hits with max 2 mismatch(es)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 37 after removing NGG 'N' mismatches (avoid double counting expanded pams)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 36 after removing NGG 'G1' mismatches(except for NGG -&gt; NAG, which is allowed)</span></div><div class='output co'>#&gt; <span class='message'>				Retain 33 after removing NGG 'G2' mismatches</span></div><div class='output co'>#&gt; <span class='message'>				Retain 33 after removing more than 2 mismatches</span></div><div class='output co'>#&gt; <span class='message'>			Count matches</span></div><div class='output co'>#&gt; <span class='message'>		Count off-targets</span></div><div class='output co'>#&gt; <span class='message'>		       30 are off0 free</span></div><div class='output co'>#&gt; <span class='message'>		       32 are off1 free</span></div><div class='output co'>#&gt; <span class='message'>		       24 are off2 free</span></div><div class='output co'>#&gt; <span class='message'>	Filter for best offtarget counts</span></div><div class='output co'>#&gt; <span class='message'>		Retain 32 with best off0 counts per pename</span></div><div class='output co'>#&gt; <span class='message'>		Retain 32 with best off1 counts per pename</span></div><div class='output co'>#&gt; <span class='message'>		Retain 24 with best off2 counts per pename</span></div><div class='output co'>#&gt; <span class='message'>	Score ontargets</span></div><div class='output co'>#&gt; <span class='message'>		Add (4-23-3) contextseqs</span></div><div class='output co'>#&gt; <span class='message'>		Score contextseqs with Doench2014</span></div><div class='input'>    <span class='fu'>plot_intervals</span>(<span class='no'>gr</span>)</div><div class='img'><img src='plot_intervals-2.png' alt='' width='700' height='433' /></div><div class='input'>    <span class='fu'>plot_intervals</span>(<span class='fu'><a href='extend_for_pe.html'>extend_for_pe</a></span>(<span class='no'>gr</span>))</div><div class='img'><img src='plot_intervals-3.png' alt='' width='700' height='433' /></div><div class='input'>    <span class='fu'>plot_intervals</span>(<span class='no'>spacers</span>)</div><div class='output co'>#&gt; <span class='warning'>Warning: Removed 34 rows containing missing values (geom_point).</span></div><div class='output co'>#&gt; <span class='warning'>Warning: Removed 34 rows containing missing values (geom_point).</span></div><div class='img'><img src='plot_intervals-4.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># Empty gr</span>
    <span class='fu'>plot_intervals</span>(<span class='kw pkg'>GenomicRanges</span><span class='kw ns'>::</span><span class='fu'><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a></span>())</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Aditya Bhagwat.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


