<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Crispr/Cas9 gRNA design • multicrispr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Crispr/Cas9 gRNA design">
<meta property="og:description" content="multicrispr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">multicrispr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.24</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/crispr_grna_design.html">Crispr/Cas9 gRNA design</a>
    </li>
    <li>
      <a href="../articles/genome_arithmetics.html">Genome arithmetics</a>
    </li>
    <li>
      <a href="../articles/prime_editing.html">Prime Editing</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Crispr/Cas9 gRNA design</h1>
                        <h4 class="author">Aditya M Bhagwat</h4>
            
            <h4 class="date">2020-09-06</h4>
      
      
      <div class="hidden name"><code>crispr_grna_design.Rmd</code></div>

    </div>

    
    
<!-- ![](../inst/extdata/readme.png) -->
<div id="background" class="section level1">
<h1 class="hasAnchor">
<a href="#background" class="anchor"></a>Background</h1>
<p><strong>Crispr/Cas9 </strong>is a two-component genome engineering tool:</p>
<ul>
<li>The <strong>Cas9</strong> enzyme performs some action at a genomic locus (cuts, activates, represses, …)</li>
<li>The <strong>guide RNA spacer</strong> targets a matching locus when followed by NGG “pam” sequence</li>
</ul>
<p><img src="../inst/extdata/crispr.png" width="594"></p>
<p><strong>Guide RNA design</strong> involves</p>
<ol style="list-style-type: decimal">
<li>Defining target ranges</li>
<li>Transforming them according to Crispr/Cas9 application</li>
<li>Finding 20 nucleotide gRNA spacers
<ul>
<li>which match the target sequence (on either strand) and are followed by an NGG “pam”</li>
<li>have <strong>minimal off-target</strong> (mis)matches</li>
<li>have <strong>maximal on-target</strong> efficiency</li>
</ul>
</li>
</ol>
</div>
<div id="multicrispr" class="section level1">
<h1 class="hasAnchor">
<a href="#multicrispr" class="anchor"></a>Multicrispr</h1>
<p>Multicrispr aism to be a gRNA design solution which is:</p>
<ul>
<li>complete: defining/transforming targets, finding spacers, analysing on/off-targets</li>
<li>performant: scales towards thousands of targets (data.table backend)</li>
<li>intuitive: a functional programming design and intuitive visualizations</li>
<li>interoperable: acceping and returning a GRanges</li>
</ul>
<p>For on-target scoring either the Doench2014 or the Doench2016 method can be used (Doench2016 is the current standard, see e.g. Haeussler et al., 2016). For off-target analysis Bowtie (fast) as well as vcountPDict (exact) can be used. The figure below gives an overview of how multicrispr can be used, the subsequent sections below discuss the details.</p>
<p><img src="../inst/extdata/overview.png" width="100%"></p>
<div id="install" class="section level2">
<h2 class="hasAnchor">
<a href="#install" class="anchor"></a>Install</h2>
<p>Installing <strong>multicrispr</strong> is straightforward:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="co"># From BioC</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)
<span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="kw">version</span><span class="kw">=</span><span class="st">'devel'</span>)
<span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"multicrispr"</span>)

<span class="co"># From gitlab: </span>
<span class="co">#url &lt;- 'https://gitlab.gwdg.de/loosolab/software/multicrispr.git'</span>
<span class="co">#remotes::install_git(url, repos = BiocManager::repositories())</span></pre></body></html></div>
<p>Doench et al. (2016) ’s python package <strong>azimuth</strong> for on-target efficiency prediction using their method can be easily installed and activated using reticulate:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="co"># Install once</span>
  <span class="co"># reticulate::conda_create('azienv', 'python=2.7')</span>
  <span class="co"># reticulate::conda_install('azienv', 'azimuth', pip = TRUE)</span>
  <span class="co"># reticulate::conda_install('azienv', 'scikit-learn==0.17.1', pip = TRUE)</span>
<span class="co"># Then activate</span>
  <span class="kw pkg">reticulate</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/reticulate/man/use_python.html">use_condaenv</a></span>(<span class="st">'azienv'</span>)</pre></body></html></div>
<p><strong>Bowtie-indexed genomes</strong> for fast offtarget analysis can be installed using <code>index_genome</code>. For the two genomes used in the examples, mm10 and hg38, the function downloads pre-build indexes from our data server, allowing a quick exploration (set <code>download=FALSE</code> to build index anew):</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="../reference/index_genome.html">index_genome</a></span>(<span class="kw pkg">BSgenome.Mmusculus.UCSC.mm10</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/BSgenome.Mmusculus.UCSC.mm10/man/package.html">BSgenome.Mmusculus.UCSC.mm10</a></span>)
<span class="fu"><a href="../reference/index_genome.html">index_genome</a></span>(<span class="kw pkg">BSgenome.Hsapiens.UCSC.hg38</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg38/man/package.html">BSgenome.Hsapiens.UCSC.hg38</a></span>  )</pre></body></html></div>
<!-- 1. Install conda for python 2.7 -->
<!-- 2. Create a new conda environment: `conda create --name azimuthenv python=2.7` -->
<!-- 3. Activate conda environment:     `conda activate azimuthenv` -->
<!-- 4. Install module azimuth:         `pip install azimuth` -->
<!-- 5. Install scikit-learn:           `pip install scikit-learn==0.17.1` -->
</div>
<div id="define-targets" class="section level2">
<h2 class="hasAnchor">
<a href="#define-targets" class="anchor"></a>Define targets</h2>
<p><strong>bed_to_granges</strong> converts a (0-based) BED coordinate file into a (1-based) GRanges.<br>
An example is loading the 1974 binding sites of the transcription factor SRF:</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="no">magrittr</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="no">multicrispr</span>)
<span class="no">bedfile</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata/SRF.bed'</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">'multicrispr'</span>)
<span class="no">targets0</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/bed_to_granges.html">bed_to_granges</a></span>(<span class="no">bedfile</span>, <span class="kw">genome</span> <span class="kw">=</span> <span class="st">'mm10'</span>)</pre></body></html></div>
<p><img src="crispr_grna_design_files/figure-html/unnamed-chunk-7-1.png" width="60%"></p>
<p><strong>genes_to_granges</strong> and <strong>genefile_to_granges</strong> convert entrez/ensembl gene identifiers into GRanges using <code>txdb</code> gene models. An example that loads a small subset of the 1974 SRF binding sites:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">entrezfile</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata/SRF.entrez'</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">'multicrispr'</span>)
<span class="no">txdb</span> <span class="kw">&lt;-</span> <span class="kw pkg">TxDb.Mmusculus.UCSC.mm10.knownGene</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/TxDb.Mmusculus.UCSC.mm10.knownGene/man/package.html">TxDb.Mmusculus.UCSC.mm10.knownGene</a></span>
<span class="fu"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="fu"><a href="../reference/genes_to_granges.html">genefile_to_granges</a></span>(<span class="no">entrezfile</span>, <span class="no">txdb</span>, <span class="kw">complement</span> <span class="kw">=</span> <span class="fl">TRUE</span>))</pre></body></html></div>
<p><img src="crispr_grna_design_files/figure-html/unnamed-chunk-8-1.png" width="60%"></p>
<p><strong>char_to_granges</strong> uses a 1-based coordinate vector to specify the GRanges.<br>
It can be used e.g. to define the first SRF binding site explicitly:</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">bsgenome</span> <span class="kw">&lt;-</span> <span class="kw pkg">BSgenome.Mmusculus.UCSC.mm10</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/BSgenome.Mmusculus.UCSC.mm10/man/package.html">BSgenome.Mmusculus.UCSC.mm10</a></span>
<span class="fu"><a href="../reference/plot_intervals.html">plot_intervals</a></span>(<span class="fu"><a href="../reference/char_to_granges.html">char_to_granges</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw">srf1</span> <span class="kw">=</span> <span class="st">'chr13:119991554-119991569:+'</span>), <span class="no">bsgenome</span>))</pre></body></html></div>
<p><img src="crispr_grna_design_files/figure-html/unnamed-chunk-9-1.png" width="50%"></p>
</div>
<div id="transform" class="section level2">
<h2 class="hasAnchor">
<a href="#transform" class="anchor"></a>Transform</h2>
<p>As a second step, the original targets may require transformation. The functions <strong>up_flank</strong> (target upstream flanks), <strong>down_flank</strong> (downstream flanks), <strong>double_flank</strong> (double flanks) and <strong>extend</strong> can be used to update the target ranges as required.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="co"># Up flank</span>
    <span class="fu"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="fu"><a href="../reference/up_flank.html">up_flank</a></span>(  <span class="no">targets0</span>, -<span class="fl">200</span>, -<span class="fl">1</span>))
<span class="co"># Down flank</span>
    <span class="fu"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="fu"><a href="../reference/up_flank.html">down_flank</a></span>( <span class="no">targets0</span>, <span class="fl">1</span>, <span class="fl">200</span>))
<span class="co"># Double flank</span>
    <span class="fu"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="fu"><a href="../reference/double_flank.html">double_flank</a></span>(<span class="no">targets0</span>, -<span class="fl">200</span>, -<span class="fl">1</span>, +<span class="fl">1</span>, +<span class="fl">200</span>))
<span class="co"># Extend</span>
    <span class="no">targets</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/up_flank.html">extend</a></span>(<span class="no">targets0</span>, -<span class="fl">22</span>, <span class="fl">22</span>, <span class="kw">plot</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<p><img src="crispr_grna_design_files/figure-html/unnamed-chunk-10-1.png" width="60%"></p>
</div>
<div id="find-spacers-count-off-targets-score-on-targets" class="section level2">
<h2 class="hasAnchor">
<a href="#find-spacers-count-off-targets-score-on-targets" class="anchor"></a>Find spacers, count off-targets, score on-targets</h2>
<p><strong>find_spacers</strong> finds N20 spacers (followed by NGG pam sites), counts off-targets, and scores on-targets. It also visualizes the results and returns them as a GRanges object with off-target counts and on-target efficiency.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="no">spacers</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/find_spacers.html">find_spacers</a></span>(
    <span class="no">targets</span>, <span class="no">bsgenome</span>, <span class="kw">complement</span><span class="kw">=</span><span class="fl">FALSE</span>, <span class="kw">mismatches</span><span class="kw">=</span><span class="fl">0</span>, <span class="kw">subtract_targets</span> <span class="kw">=</span> <span class="fl">TRUE</span>)</pre></body></html></div>
<p><img src="crispr_grna_design_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
</div>
<div id="returnwrite" class="section level2">
<h2 class="hasAnchor">
<a href="#returnwrite" class="anchor"></a>Return/Write</h2>
<p>The results are returned as a GRanges object. They can be written to file with <strong>write_ranges</strong><br>
They can be cast into a data.table with <strong>gr2dt</strong>, as shown below</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r">  <span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span>(<span class="fu"><a href="../reference/gr2dt.html">gr2dt</a></span>(<span class="no">spacers</span>), <span class="kw">vec.len</span><span class="kw">=</span><span class="fl">2</span>)
<span class="co">## Classes 'data.table' and 'data.frame':   6616 obs. of  19 variables:</span>
<span class="co">##  $ seqnames    : Factor w/ 66 levels "chr1","chr2",..: 1 1 1 1 1 ...</span>
<span class="co">##  $ start       : int  4712619 4712620 4712627 4712633 4712634 ...</span>
<span class="co">##  $ end         : int  4712638 4712639 4712646 4712652 4712653 ...</span>
<span class="co">##  $ width       : int  20 20 20 20 20 ...</span>
<span class="co">##  $ strand      : Factor w/ 3 levels "+","-","*": 2 2 2 2 2 ...</span>
<span class="co">##  $ targetname  : chr  "chr1:4712628-4712643:-" "chr1:4712628-4712643:-" ...</span>
<span class="co">##  $ targetstart : int  4712628 4712628 4712628 4712628 4712628 ...</span>
<span class="co">##  $ targetend   : int  4712643 4712643 4712643 4712643 4712643 ...</span>
<span class="co">##  $ crisprname  : chr  "chr1:4712628-4712643:-_5" "chr1:4712628-4712643:-_4" ...</span>
<span class="co">##  $ crisprspacer: chr  "ATATAAGGGCATTGGAAGAA" "AATATAAGGGCATTGGAAGA" ...</span>
<span class="co">##  $ crisprpam   : chr  "GGG" "AGG" ...</span>
<span class="co">##  $ name        : chr  "SRF_MA0083.3" "SRF_MA0083.3" ...</span>
<span class="co">##  $ score       : num  10.5 10.5 ...</span>
<span class="co">##  $ G0          : int  7 7 7 8 1678 ...</span>
<span class="co">##  $ T0          : int  1 1 1 2 825 ...</span>
<span class="co">##  $ off         : int  6 6 6 6 853 ...</span>
<span class="co">##  $ off0        : int  6 6 6 6 853 ...</span>
<span class="co">##  $ Doench2014  : num  0.16491 0.00721 ...</span>
<span class="co">##  $ names       : chr  "chr1:4712628-4712643:-_5" "chr1:4712628-4712643:-_4" ...</span>
<span class="co">##  - attr(*, ".internal.selfref")=&lt;externalptr&gt;</span></pre></body></html></div>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>Anzalone, A.V., Randolph, P.B., Davis, J.R. et al. Search-and-replace genome editing without double-strand breaks or donor DNA. Nature 576, 149–157 (2019). <a href="https://doi.org/10.1038/s41586-019-1711-4" class="uri">https://doi.org/10.1038/s41586-019-1711-4</a></p>
<p>Doench JG, Hartenian E, Graham DB, Tothova Z, Hegde M, Smith I, Sullender M, Ebert BL, Xavier RJ, Root DE, Rational design of highly active sgRNAs for CRISPR-Cas9-mediated gene inactivation Nat Biotechnol 2014 1262-1267.</p>
<p>Doench JG, Fusi N, Sullender M, Hegde M, Vaimberg EW, Donovan KF, Smith I, Tothova Z, Wilen C, Orchard R, Optimized sgRNA design to maximize activity and minimize off-target effects of CRISPR-Cas9 Nat Biotechnol 2016 184-191</p>
<p>Haeussler M, Schönig K, Eckert H, Eschstruth A, Mianné J, Renaud J-B, Schneider-Maunoury S, Shkumatava A, Teboul L, Kent J, Evaluation of off-target and on-target scoring algorithms and integration into the guide RNA selection tool CRISPOR Genome Biol 2016 148</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Aditya Bhagwat.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
