<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using multicrispr for gRNA design • multicrispr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using multicrispr for gRNA design">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">multicrispr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/designing_grna_libraries.html">Using multicrispr for gRNA design</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Using multicrispr for gRNA design</h1>
                        <h4 class="author">Aditya M Bhagwat</h4>
            
            <h4 class="date">2020-04-23</h4>
      
      
      <div class="hidden name"><code>designing_grna_libraries.Rmd</code></div>

    </div>

    
    
<!-- ![](../inst/extdata/readme.png) -->
<div id="background" class="section level1">
<h1 class="hasAnchor">
<a href="#background" class="anchor"></a>Background</h1>
<div id="crisprcas9-and-prime-editing" class="section level2">
<h2 class="hasAnchor">
<a href="#crisprcas9-and-prime-editing" class="anchor"></a>Crispr/Cas9 and Prime Editing</h2>
<p><strong>Crispr/Cas9</strong> is a prokaryotic immune system turned into a genome engineering tool, with a Cas9/gRNA complex at the heart of its functioning. The <strong>gRNA</strong> (guide RNA) contains a 20 nucleotide (proto)<strong>spacer</strong> which guides the complex to a genomic locus with identical sequence, but only when followed by an <em>NGG</em> <strong>PAM</strong> (protospacer adjacent motif (see Figure 1A). The <strong>Cas9</strong> enzyme performs some effector action at that locus: <em>wildtype Cas9</em> cuts both strands after spacer nucleotide 17, <em>Cas9Nickase</em> cuts only a single strand (two variants exist, cutting respectively each of the strands), while catalytically dead <em>dCas9</em>, fused to some effector domain (KRAB, VPR, APEX) performs some alternative action (e.g. KRAB represses, VPR activates, APEX biotinylates).</p>
<p><strong>Prime Editing</strong> extends this further, by coupling Cas9 to a Reverse Transcriptase (RT). As shown in Figure 1, the guide RNA is extended with a 3’ extension that contains primer binding site (binding a complementary DNA region that acts as a primer for reverse transcription) and reverse transcription template (the template used for reverse transcription). The spacer still acts as targeting agent, guiding the Cas9/RT complex to targeted genomic locus.</p>
<div class="figure">
<img src="../inst/extdata/pe.png" alt="gRNAs for Crispr/Cas9 (A) and Prime Editing" width="352"><p class="caption">
gRNAs for Crispr/Cas9 (A) and Prime Editing
</p>
</div>
</div>
<div id="guide-rna-design" class="section level2">
<h2 class="hasAnchor">
<a href="#guide-rna-design" class="anchor"></a>Guide RNA design</h2>
<p><strong>Guide RNA design</strong> involves finding good guide RNAs to hit the targets of interest. Plasmids with guide RNA scaffolds, ready for cloning desired spacers into, can be readily ordered from AddGene. The actual task, therefore, boils down to finding good spacers for the targets of interest. For prime editing, it additionally involves finding appropriate primer binding site and reverse transcription template to edit the target site of interest.</p>
<p>A <strong>good spacer</strong> needs to fulfill two requirements.</p>
<ol style="list-style-type: decimal">
<li>
<p><strong>Minimal off-target</strong> (mis)matches, so that only intended targets are hit.</p>
<ul>
<li>Crispr gRNAs can hit exact (with identical and alternate NGG pam), as well as (up-to 2) mismatch offtargets.</li>
<li>Prime Editing is much more specific, hitting only exact offtargets.</li>
</ul>
</li>
<li><p><strong>Maximal on-target</strong> efficiency. Over the years, several sequence-to-efficiency prediction models have been developed, of which the Doench 2016 score has currently become the community standard. The score is not perfect, and examples can be found where the prediction and the actual outcome differ. Yet, to-date it is the best heuristic to predict on-target efficiency, worth making use of.</p></li>
</ol>
</div>
</div>
<div id="multicrispr" class="section level1">
<h1 class="hasAnchor">
<a href="#multicrispr" class="anchor"></a>Multicrispr</h1>
<p>Multicrispr was developed to make guide RNA design easier. As shown below, it contains functions to <strong>define</strong> and <strong>transform</strong> targets, <strong>find spacers</strong>, compute/add <strong>offtarget counts</strong> and <strong>efficiency scores</strong> (Doench 2016), and finally return all of this as a <strong>GRanges</strong> object.</p>
<p><img src="../inst/extdata/readme_portrait.png" width="80%"></p>
<div id="install" class="section level2">
<h2 class="hasAnchor">
<a href="#install" class="anchor"></a>Install</h2>
<p>Installing <strong>multicrispr</strong> is simple:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1">url &lt;-<span class="st"> 'https://gitlab.gwdg.de/loosolab/software/multicrispr.git'</span></a>
<a class="sourceLine" id="cb1-2" title="2">remotes<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/remotes/man/install_git.html">install_git</a></span>(url, <span class="dt">repos =</span> BiocManager<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/BiocManager/man/repositories.html">repositories</a></span>())</a></code></pre></div>
<p>Doench et al. (2016) ’s python package <strong>azimuth</strong> for on-target efficiency prediction using their method can be easily installed and activated using reticulate:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="co"># Install - run R(Studio) with admin privileges for this to work!</span></a>
<a class="sourceLine" id="cb2-2" title="2">  reticulate<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/reticulate/man/conda-tools.html">conda_create</a></span>(<span class="st">'azienv'</span>, <span class="st">'python=2.7'</span>)</a>
<a class="sourceLine" id="cb2-3" title="3">  reticulate<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/reticulate/man/conda-tools.html">conda_install</a></span>(<span class="st">'azienv'</span>, <span class="st">'azimuth'</span>, <span class="dt">pip =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb2-4" title="4">  reticulate<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/reticulate/man/conda-tools.html">conda_install</a></span>(<span class="st">'azienv'</span>, <span class="st">'scikit-learn==0.17.1'</span>, <span class="dt">pip =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="co"># Activate</span></a>
<a class="sourceLine" id="cb3-2" title="2">  reticulate<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/reticulate/man/use_python.html">use_condaenv</a></span>(<span class="st">'azienv'</span>)</a></code></pre></div>
<p>Bowtie-indexed genomes for quick offtarget analysis can be installed using <code>index_genome</code>. For the two genomes used in the examples, mm10 and hg38, the functions downloads pre-build indexes from our data server, allowing a quick exploration (set <code>download=FALSE</code> to build index anew):</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="kw"><a href="../reference/index_genome.html">index_genome</a></span>(BSgenome.Mmusculus.UCSC.mm10<span class="op">::</span>BSgenome.Mmusculus.UCSC.mm10)</a>
<a class="sourceLine" id="cb4-2" title="2"><span class="kw"><a href="../reference/index_genome.html">index_genome</a></span>(BSgenome.Hsapiens.UCSC.hg38<span class="op">::</span>BSgenome.Hsapiens.UCSC.hg38  )</a></code></pre></div>
<!-- 1. Install conda for python 2.7 -->
<!-- 2. Create a new conda environment: `conda create --name azimuthenv python=2.7` -->
<!-- 3. Activate conda environment:     `conda activate azimuthenv` -->
<!-- 4. Install module azimuth:         `pip install azimuth` -->
<!-- 5. Install scikit-learn:           `pip install scikit-learn==0.17.1` -->
</div>
<div id="define-targets" class="section level2">
<h2 class="hasAnchor">
<a href="#define-targets" class="anchor"></a>Define targets</h2>
<p><img src="../inst/extdata/01_define_targets.png" width="70%"></p>
<p><code>bed_to_granges</code> converts a (0-based) BED coordinate file into a (1-based) GRanges.<br>
An example is loading the 1974 binding sites of the transcription factor SRF:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(magrittr)</a>
<a class="sourceLine" id="cb5-2" title="2"><span class="co">## Lade nötiges Paket: magrittr</span></a>
<a class="sourceLine" id="cb5-3" title="3"><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(multicrispr)</a>
<a class="sourceLine" id="cb5-4" title="4"><span class="co">## Lade nötiges Paket: multicrispr</span></a>
<a class="sourceLine" id="cb5-5" title="5">bedfile &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata/SRF.bed'</span>, <span class="dt">package =</span> <span class="st">'multicrispr'</span>)</a>
<a class="sourceLine" id="cb5-6" title="6">tfbs0 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_to_granges.html">bed_to_granges</a></span>(bedfile, <span class="dt">genome =</span> <span class="st">'mm10'</span>)</a>
<a class="sourceLine" id="cb5-7" title="7"><span class="co">##  Read SRF.bed into GRanges</span></a>
<a class="sourceLine" id="cb5-8" title="8"><span class="co">##      1974 ranges on 21 chromosomes</span></a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p><code>genes_to_granges</code> and <code>genefile_to_granges</code> convert entrez/ensembl gene identifiers into GRanges using <code>txdb</code> gene models. An example that loads of a small subset of the 1974 SRF binding sites:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(multicrispr)</a>
<a class="sourceLine" id="cb6-2" title="2">entrezfile &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata/SRF.entrez'</span>, <span class="dt">package =</span> <span class="st">'multicrispr'</span>)</a>
<a class="sourceLine" id="cb6-3" title="3">txdb &lt;-<span class="st"> </span>TxDb.Mmusculus.UCSC.mm10.knownGene<span class="op">::</span>TxDb.Mmusculus.UCSC.mm10.knownGene</a>
<a class="sourceLine" id="cb6-4" title="4">sometfbs0 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/genes_to_granges.html">genefile_to_granges</a></span>(entrezfile, txdb, <span class="dt">complement =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb6-5" title="5"><span class="co">##      Convert 10 genes to 10 GRanges</span></a>
<a class="sourceLine" id="cb6-6" title="6"><span class="co">##      20 ranges after adding inverse strands</span></a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p><code>char_to_granges</code> uses a 1-based coordinate vector to specify the GRanges.<br>
It can be used to define four prime editing sites mentioned by Anzalone et al. (2019):</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1"><span class="co"># char_to_granges: Anzalone et al. (2019) prime editing targets</span></a>
<a class="sourceLine" id="cb7-2" title="2">bsgenome &lt;-<span class="st"> </span>BSgenome.Hsapiens.UCSC.hg38<span class="op">::</span>BSgenome.Hsapiens.UCSC.hg38  </a>
<a class="sourceLine" id="cb7-3" title="3">pe0 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/char_to_granges.html">char_to_granges</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dt">HBB  =</span> <span class="st">'chr11:5227002:-'</span>), bsgenome)</a>
<a class="sourceLine" id="cb7-4" title="4"><span class="kw"><a href="../reference/plot_intervals.html">plot_intervals</a></span>(pe0)</a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-11-1.png" width="336"></p>
</div>
<div id="transform-targets" class="section level2">
<h2 class="hasAnchor">
<a href="#transform-targets" class="anchor"></a>Transform targets</h2>
<p><img src="../inst/extdata/02_transform.png" width="70%"></p>
<p>As a second step, the original targets may require transformation. The functions <code>extend</code>, <code>up_flank</code> (upstream flanks), <code>down_flank</code> (downstream flanks), and <code>double_flank</code> (double flanks) can be used to update the target ranges as required.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="co"># Extend</span></a>
<a class="sourceLine" id="cb8-2" title="2">targets &lt;-<span class="st"> </span>pe0</a>
<a class="sourceLine" id="cb8-3" title="3"><span class="kw"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="kw"><a href="../reference/up_flank.html">extend</a></span>(targets, <span class="dv">-22</span>, <span class="dv">22</span>, <span class="dt">plot =</span> <span class="ot">TRUE</span>))</a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-13-1.png" width="65%"></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1"><span class="co"># Up flank</span></a>
<a class="sourceLine" id="cb9-2" title="2"><span class="kw"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="kw"><a href="../reference/up_flank.html">up_flank</a></span>(  targets, <span class="dv">-200</span>, <span class="dv">-1</span>, <span class="dt">plot =</span> <span class="ot">TRUE</span>))</a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-13-2.png" width="65%"></p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1"><span class="co"># Down flank</span></a>
<a class="sourceLine" id="cb10-2" title="2"><span class="kw"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="kw"><a href="../reference/up_flank.html">down_flank</a></span>( targets, <span class="dv">1</span>, <span class="dv">200</span>, <span class="dt">plot =</span> <span class="ot">TRUE</span>))</a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-13-3.png" width="65%"></p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1"><span class="co"># Double flank</span></a>
<a class="sourceLine" id="cb11-2" title="2"><span class="kw"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="kw"><a href="../reference/double_flank.html">double_flank</a></span>(targets, <span class="dv">-200</span>, <span class="dv">-1</span>, <span class="op">+</span><span class="dv">1</span>, <span class="op">+</span><span class="dv">200</span>, <span class="dt">plot =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb11-3" title="3"><span class="co">##      2 flank ranges: 1 up + 1 down</span></a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-13-4.png" width="65%"></p>
</div>
<div id="find-spacers" class="section level2">
<h2 class="hasAnchor">
<a href="#find-spacers" class="anchor"></a>Find spacers</h2>
<p><img src="../inst/extdata/03_find.png" width="70%"></p>
<p>The next step is to find N20 spacers followed by NGG pam sites. This can be done with <code>find_spacers</code>, after extending target range(s) to prime editing neighbourhood(s):</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1">bsgenome   &lt;-<span class="st"> </span>BSgenome.Hsapiens.UCSC.hg38<span class="op">::</span>BSgenome.Hsapiens.UCSC.hg38</a>
<a class="sourceLine" id="cb12-2" title="2">neighbourhood &lt;-<span class="st"> </span><span class="kw"><a href="../reference/extend_for_pe.html">extend_for_pe</a></span>(targets, bsgenome)</a>
<a class="sourceLine" id="cb12-3" title="3">spacers    &lt;-<span class="st"> </span><span class="kw"><a href="../reference/find_spacers.html">find_spacers</a></span>(neighbourhood, bsgenome, <span class="dt">complement =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-15-1.png" width="288"></p>
<p>For prime editing, however, it is better to use <code>find_pe_spacers</code>, which performs target to neighbourhood extension implicitly, and also adds 3’ extension (Fig 1) to returned spacer GRanges.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1">spacers &lt;-<span class="st"> </span><span class="kw"><a href="../reference/find_pe_spacers.html">find_pe_spacers</a></span>(targets, <span class="dt">bsgenome=</span>bsgenome) </a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-16-1.png" width="316.8"></p>
<p>This finds prime editing sites that follow Anzalone et al. (2019)’s recommended default settings: a 13-nucleotide primer and a 16-nucleotide reverse transcription template. These settings can be easily altered (e.g. Anzalone et al. report success with reverse transcription templates of up to 48 nucleotides) by:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1">spacers &lt;-<span class="st"> </span><span class="kw"><a href="../reference/find_pe_spacers.html">find_pe_spacers</a></span>(targets, <span class="dt">bsgenome=</span>bsgenome, <span class="dt">nrt =</span> <span class="dv">48</span>) </a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-17-1.png" width="345.6"></p>
</div>
<div id="add-offtargets" class="section level2">
<h2 class="hasAnchor">
<a href="#add-offtargets" class="anchor"></a>Add offtargets</h2>
<p><img src="../inst/extdata/04_offtargets.png" width="70%"></p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1"><span class="cf">if</span> (<span class="kw"><a href="../reference/has_been_indexed.html">has_been_indexed</a></span>(bsgenome)){</a>
<a class="sourceLine" id="cb15-2" title="2">  spacers <span class="op">%&lt;&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/add_offtargets.html">add_offtargets</a></span>(bsgenome, <span class="dt">mismatch =</span> <span class="dv">0</span>, <span class="dt">plot =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb15-3" title="3">}</a>
<a class="sourceLine" id="cb15-4" title="4"><span class="co">##  Add genome counts</span></a>
<a class="sourceLine" id="cb15-5" title="5"><span class="co">##      Expand iupac ambiguities in pam</span></a>
<a class="sourceLine" id="cb15-6" title="6"><span class="co">##      Write reads to ~/multicrisprout/spacers.fa</span></a>
<a class="sourceLine" id="cb15-7" title="7"><span class="co">##      Map reads: ~/multicrisprout/spacers/spacers_to_BSgenome.Hsapiens.UCSC.hg38.txt</span></a>
<a class="sourceLine" id="cb15-8" title="8"><span class="co">##      Load results</span></a>
<a class="sourceLine" id="cb15-9" title="9"><span class="co">##      Count matches</span></a>
<a class="sourceLine" id="cb15-10" title="10"><span class="co">##  Calculate offtargets for 6 spacers</span></a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-19-1.png" width="345.6"></p>
</div>
<div id="add-efficiency" class="section level2">
<h2 class="hasAnchor">
<a href="#add-efficiency" class="anchor"></a>Add efficiency</h2>
<p><img src="../inst/extdata/05_efficiency.png" width="70%"></p>
<p>Not all N<sub>20</sub>NGG gRNA sequences target equally well (even when matching sequence perfectly). For each position in the 23-bp gRNA sequence, the nucleotide present in current, previous and next position has an effect on targeting efficiency. The current community standard for on-target efficiency prediction is that of Doench et al. (2016), which can be computed and added with <code>add_efficiency</code>:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1"><span class="cf">if</span> (reticulate<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/reticulate/man/py_module_available.html">py_module_available</a></span>(<span class="st">'azimuth'</span>)){</a>
<a class="sourceLine" id="cb16-2" title="2">  spacers <span class="op">%&lt;&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/add_efficiency.html">add_efficiency</a></span>(bsgenome, <span class="st">'Doench2016'</span>)</a>
<a class="sourceLine" id="cb16-3" title="3">}</a>
<a class="sourceLine" id="cb16-4" title="4"><span class="co">##  Score crispr spacers</span></a>
<a class="sourceLine" id="cb16-5" title="5"><span class="co">##      Add (4-23-3) contextseqs</span></a>
<a class="sourceLine" id="cb16-6" title="6"><span class="co">##      Score contextseqs with Doench2016 (azimuth)</span></a>
<a class="sourceLine" id="cb16-7" title="7"><span class="co">##      Run Doench2016 1 times on 10000-seq chunks to preserve memory</span></a>
<a class="sourceLine" id="cb16-8" title="8"><span class="co">## No model file specified, using V3_model_nopos</span></a>
<a class="sourceLine" id="cb16-9" title="9"><span class="co">##      Completed in 1.1 secs</span></a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-21-1.png" width="345.6"></p>
</div>
<div id="return-spacer-granges" class="section level2">
<h2 class="hasAnchor">
<a href="#return-spacer-granges" class="anchor"></a>Return spacer GRanges</h2>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1">  spacers</a>
<a class="sourceLine" id="cb17-2" title="2"><span class="co">## GRanges object with 6 ranges and 14 metadata columns:</span></a>
<a class="sourceLine" id="cb17-3" title="3"><span class="co">##         seqnames          ranges strand |                  crisprcontext</span></a>
<a class="sourceLine" id="cb17-4" title="4"><span class="co">##            &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |                    &lt;character&gt;</span></a>
<a class="sourceLine" id="cb17-5" title="5"><span class="co">##   HBB_1    chr11 5226958-5226977      + | TCATCCACGTTCACCTTGCCCCACAGGGCA</span></a>
<a class="sourceLine" id="cb17-6" title="6"><span class="co">##   HBB_2    chr11 5226959-5226978      + | CATCCACGTTCACCTTGCCCCACAGGGCAG</span></a>
<a class="sourceLine" id="cb17-7" title="7"><span class="co">##   HBB_3    chr11 5226968-5226987      + | TCACCTTGCCCCACAGGGCAGTAACGGCAG</span></a>
<a class="sourceLine" id="cb17-8" title="8"><span class="co">##   HBB_4    chr11 5227003-5227022      - | ACACCATGGTGCATCTGACTCCTGAGGAGA</span></a>
<a class="sourceLine" id="cb17-9" title="9"><span class="co">##   HBB_5    chr11 5226984-5227003      + | GGCAGTAACGGCAGACTTCTCCTCAGGAGT</span></a>
<a class="sourceLine" id="cb17-10" title="10"><span class="co">##   HBB_6    chr11 5227021-5227040      - | ACTAGCAACCTCAAACAGACACCATGGTGC</span></a>
<a class="sourceLine" id="cb17-11" title="11"><span class="co">##          targetname targetstart targetend  crisprname         crisprspacer</span></a>
<a class="sourceLine" id="cb17-12" title="12"><span class="co">##         &lt;character&gt;   &lt;integer&gt; &lt;integer&gt; &lt;character&gt;          &lt;character&gt;</span></a>
<a class="sourceLine" id="cb17-13" title="13"><span class="co">##   HBB_1         HBB     5227002   5227002       HBB_1 CCACGTTCACCTTGCCCCAC</span></a>
<a class="sourceLine" id="cb17-14" title="14"><span class="co">##   HBB_2         HBB     5227002   5227002       HBB_2 CACGTTCACCTTGCCCCACA</span></a>
<a class="sourceLine" id="cb17-15" title="15"><span class="co">##   HBB_3         HBB     5227002   5227002       HBB_3 CTTGCCCCACAGGGCAGTAA</span></a>
<a class="sourceLine" id="cb17-16" title="16"><span class="co">##   HBB_4         HBB     5227002   5227002       HBB_4 CATGGTGCATCTGACTCCTG</span></a>
<a class="sourceLine" id="cb17-17" title="17"><span class="co">##   HBB_5         HBB     5227002   5227002       HBB_5 GTAACGGCAGACTTCTCCTC</span></a>
<a class="sourceLine" id="cb17-18" title="18"><span class="co">##   HBB_6         HBB     5227002   5227002       HBB_6 GCAACCTCAAACAGACACCA</span></a>
<a class="sourceLine" id="cb17-19" title="19"><span class="co">##           crisprpam        primer</span></a>
<a class="sourceLine" id="cb17-20" title="20"><span class="co">##         &lt;character&gt;   &lt;character&gt;</span></a>
<a class="sourceLine" id="cb17-21" title="21"><span class="co">##   HBB_1         AGG GTTCACCTTGCCC</span></a>
<a class="sourceLine" id="cb17-22" title="22"><span class="co">##   HBB_2         GGG TTCACCTTGCCCC</span></a>
<a class="sourceLine" id="cb17-23" title="23"><span class="co">##   HBB_3         CGG CCCCACAGGGCAG</span></a>
<a class="sourceLine" id="cb17-24" title="24"><span class="co">##   HBB_4         AGG GTGCATCTGACTC</span></a>
<a class="sourceLine" id="cb17-25" title="25"><span class="co">##   HBB_5         AGG CGGCAGACTTCTC</span></a>
<a class="sourceLine" id="cb17-26" title="26"><span class="co">##   HBB_6         TGG CCTCAAACAGACA</span></a>
<a class="sourceLine" id="cb17-27" title="27"><span class="co">##                                            revtranscript</span></a>
<a class="sourceLine" id="cb17-28" title="28"><span class="co">##                                              &lt;character&gt;</span></a>
<a class="sourceLine" id="cb17-29" title="29"><span class="co">##   HBB_1 CACAGGGCAGTAACGGCAGACTTCTCCTCAGGAGTCAGATGCACCATG</span></a>
<a class="sourceLine" id="cb17-30" title="30"><span class="co">##   HBB_2 ACAGGGCAGTAACGGCAGACTTCTCCTCAGGAGTCAGATGCACCATGG</span></a>
<a class="sourceLine" id="cb17-31" title="31"><span class="co">##   HBB_3 TAACGGCAGACTTCTCCTCAGGAGTCAGATGCACCATGGTGTCTGTTT</span></a>
<a class="sourceLine" id="cb17-32" title="32"><span class="co">##   HBB_4 CTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGG</span></a>
<a class="sourceLine" id="cb17-33" title="33"><span class="co">##   HBB_5 CTCAGGAGTCAGATGCACCATGGTGTCTGTTTGAGGTTGCTAGTGAAC</span></a>
<a class="sourceLine" id="cb17-34" title="34"><span class="co">##   HBB_6 CCATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGT</span></a>
<a class="sourceLine" id="cb17-35" title="35"><span class="co">##                                                             extension        G0</span></a>
<a class="sourceLine" id="cb17-36" title="36"><span class="co">##                                                           &lt;character&gt; &lt;integer&gt;</span></a>
<a class="sourceLine" id="cb17-37" title="37"><span class="co">##   HBB_1 CATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAAC         1</span></a>
<a class="sourceLine" id="cb17-38" title="38"><span class="co">##   HBB_2 CCATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAA         1</span></a>
<a class="sourceLine" id="cb17-39" title="39"><span class="co">##   HBB_3 AAACAGACACCATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG         1</span></a>
<a class="sourceLine" id="cb17-40" title="40"><span class="co">##   HBB_4 CCACGTTCACCTTGCCCCACAGGGCAGTAACGGCAGACTTCTCCTCAGGAGTCAGATGCAC         2</span></a>
<a class="sourceLine" id="cb17-41" title="41"><span class="co">##   HBB_5 GTTCACTAGCAACCTCAAACAGACACCATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCG         1</span></a>
<a class="sourceLine" id="cb17-42" title="42"><span class="co">##   HBB_6 ACAGGGCAGTAACGGCAGACTTCTCCTCAGGAGTCAGATGCACCATGGTGTCTGTTTGAGG         2</span></a>
<a class="sourceLine" id="cb17-43" title="43"><span class="co">##               off      off0        Doench2016</span></a>
<a class="sourceLine" id="cb17-44" title="44"><span class="co">##         &lt;numeric&gt; &lt;numeric&gt;         &lt;numeric&gt;</span></a>
<a class="sourceLine" id="cb17-45" title="45"><span class="co">##   HBB_1         0         0 0.462600534483403</span></a>
<a class="sourceLine" id="cb17-46" title="46"><span class="co">##   HBB_2         0         0 0.611140169152868</span></a>
<a class="sourceLine" id="cb17-47" title="47"><span class="co">##   HBB_3         0         0 0.485226524206282</span></a>
<a class="sourceLine" id="cb17-48" title="48"><span class="co">##   HBB_4         1         1 0.666241339730174</span></a>
<a class="sourceLine" id="cb17-49" title="49"><span class="co">##   HBB_5         0         0 0.517610901468318</span></a>
<a class="sourceLine" id="cb17-50" title="50"><span class="co">##   HBB_6         1         1  0.68921726226521</span></a>
<a class="sourceLine" id="cb17-51" title="51"><span class="co">##   -------</span></a>
<a class="sourceLine" id="cb17-52" title="52"><span class="co">##   seqinfo: 455 sequences (1 circular) from hg38 genome</span></a></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#background">Background</a><ul class="nav nav-pills nav-stacked">
<li><a href="#crisprcas9-and-prime-editing">Crispr/Cas9 and Prime Editing</a></li>
      <li><a href="#guide-rna-design">Guide RNA design</a></li>
      </ul>
</li>
      <li>
<a href="#multicrispr">Multicrispr</a><ul class="nav nav-pills nav-stacked">
<li><a href="#install">Install</a></li>
      <li><a href="#define-targets">Define targets</a></li>
      <li><a href="#transform-targets">Transform targets</a></li>
      <li><a href="#find-spacers">Find spacers</a></li>
      <li><a href="#add-offtargets">Add offtargets</a></li>
      <li><a href="#add-efficiency">Add efficiency</a></li>
      <li><a href="#return-spacer-granges">Return spacer GRanges</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Aditya Bhagwat.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
