<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>multicrispr: gRNA design for Crispr/Cas9 and Prime Editing • multicrispr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="multicrispr: gRNA design for Crispr/Cas9 and Prime Editing">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">multicrispr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/designing_grna_libraries.html">multicrispr: gRNA design for Crispr/Cas9 and Prime Editing</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>multicrispr: gRNA design for Crispr/Cas9 and Prime Editing</h1>
                        <h4 class="author">Aditya M Bhagwat</h4>
            
            <h4 class="date">2020-04-20</h4>
      
      
      <div class="hidden name"><code>designing_grna_libraries.Rmd</code></div>

    </div>

    
    
<!-- ![](../inst/extdata/readme.png) -->
<div id="background" class="section level1">
<h1 class="hasAnchor">
<a href="#background" class="anchor"></a>Background</h1>
<div id="crisprcas9-and-prime-editing" class="section level2">
<h2 class="hasAnchor">
<a href="#crisprcas9-and-prime-editing" class="anchor"></a>Crispr/Cas9 and Prime Editing</h2>
<p><strong>Crispr/Cas9</strong> is a prokaryotic immune system turned into a genome engineering tool, with a Cas9/gRNA complex at the heart of its functioning. The <strong>gRNA</strong> (guide RNA) contains a 20 nucleotide (proto)<strong>spacer</strong> which guides the complex to a genomic locus with identical sequence, but only when followed by an <em>NGG</em> <strong>PAM</strong> (protospacer adjacent motif (see Figure 1A). The <strong>Cas9</strong> enzyme performs some effector action at that locus: <em>wildtype Cas9</em> cuts both strands after spacer nucleotide 17, <em>Cas9Nickase</em> cuts only a single strand (two variants exist, cutting respectively each of the strands), while catalytically dead <em>dCas9</em>, fused to some effector domain (KRAB, VPR, APEX) performs some alternative action (e.g. KRAB represses, VPR activates, APEX biotinylates).</p>
<p><strong>Prime Editing</strong> extends this further, by coupling Cas9 to a Reverse Transcriptase (RT). As shown in Figure 1, the guide RNA is extended with a 3’ extension that contains primer binding site (binding a complementary DNA region that acts as a primer for reverse transcription) and reverse transcription template (the template used for reverse transcription). The spacer still acts as targeting agent, guiding the Cas9/RT complex to targeted genomic locus.</p>
<div class="figure">
<img src="../inst/extdata/pe.png" alt="gRNAs for Crispr/Cas9 (A) and Prime Editing" width="352"><p class="caption">
gRNAs for Crispr/Cas9 (A) and Prime Editing
</p>
</div>
</div>
<div id="guide-rna-design" class="section level2">
<h2 class="hasAnchor">
<a href="#guide-rna-design" class="anchor"></a>Guide RNA design</h2>
<p><strong>Guide RNA design</strong> involves finding good guide RNAs to hit the targets of interest. Plasmids with guide RNA scaffolds, ready for cloning desired spacers into, can be readily ordered from AddGene. The actual task, therefore, boils down to finding good spacers for the targets of interest. For prime editing, it additionally involves finding appropriate primer binding site and reverse transcription template to edit the target site of interest.</p>
<p>A <strong>good spacer</strong> needs to fulfill two requirements.</p>
<ol style="list-style-type: decimal">
<li>
<p><strong>Minimal off-target</strong> (mis)matches, so that only intended targets are hit.</p>
<ul>
<li>Crispr gRNAs can hit exact (with identical and alternate NGG pam), as well as (up-to 2) mismatch offtargets.</li>
<li>Prime Editing is much more specific, hitting only exact offtargets.</li>
</ul>
</li>
<li><p><strong>Maximal on-target</strong> efficiency. Over the years, several sequence-to-efficiency prediction models have been developed, of which the Doench 2016 score has currently become the community standard. The score is not perfect, and examples can be found where the prediction and the actual outcome differ. Yet, to-date it is the best heuristic to predict on-target efficiency, worth making use of.</p></li>
</ol>
</div>
</div>
<div id="multicrispr" class="section level1">
<h1 class="hasAnchor">
<a href="#multicrispr" class="anchor"></a>Multicrispr</h1>
<p>Multicrispr was developed to make guide RNA design easier. As shown below, it contains functions to <strong>define</strong> and <strong>transform</strong> targets, <strong>find spacers</strong>, compute/add <strong>specificity</strong> (absence of offtargets) and <strong>efficiency</strong> (Doench 2016), and finally return all of this as a <strong>GRanges</strong> object.</p>
<p><img src="../inst/extdata/readme_portrait.png" width="80%"></p>
<div id="install" class="section level2">
<h2 class="hasAnchor">
<a href="#install" class="anchor"></a>Install</h2>
<p>Installing <strong>multicrispr</strong> is simple:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1">url &lt;-<span class="st"> 'https://gitlab.gwdg.de/loosolab/software/multicrispr.git'</span></a>
<a class="sourceLine" id="cb1-2" title="2">remotes<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/remotes/man/install_git.html">install_git</a></span>(url, <span class="dt">repos =</span> BiocManager<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/BiocManager/man/repositories.html">repositories</a></span>())</a></code></pre></div>
<p>Doench et al. (2016) ’s python package <strong>azimuth</strong> for on-target efficiency prediction using their method can be easily installed and activated using reticulate:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="co"># Install - run R(Studio) with admin privileges for this to work!</span></a>
<a class="sourceLine" id="cb2-2" title="2">  reticulate<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/reticulate/man/conda-tools.html">conda_create</a></span>(<span class="st">'azienv'</span>, <span class="st">'python=2.7'</span>)</a>
<a class="sourceLine" id="cb2-3" title="3">  reticulate<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/reticulate/man/conda-tools.html">conda_install</a></span>(<span class="st">'azienv'</span>, <span class="st">'azimuth'</span>, <span class="dt">pip =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb2-4" title="4">  reticulate<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/reticulate/man/conda-tools.html">conda_install</a></span>(<span class="st">'azienv'</span>, <span class="st">'scikit-learn==0.17.1'</span>, <span class="dt">pip =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"><span class="co"># Activate</span></a>
<a class="sourceLine" id="cb3-2" title="2">  reticulate<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/reticulate/man/use_python.html">use_condaenv</a></span>(<span class="st">'azienv'</span>)</a></code></pre></div>
<p>Bowtie-indexed genomes for quick offtarget analysis can be installed using <code>index_genome</code>. For the two genomes used in the examples, mm10 and hg38, the functions downloads pre-build indexes from our data server, allowing a quick exploration (set <code>download=FALSE</code> to build index anew):</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1"><span class="kw"><a href="../reference/index_genome.html">index_genome</a></span>(BSgenome.Mmusculus.UCSC.mm10<span class="op">::</span>BSgenome.Mmusculus.UCSC.mm10)</a>
<a class="sourceLine" id="cb4-2" title="2"><span class="kw"><a href="../reference/index_genome.html">index_genome</a></span>(BSgenome.Hsapiens.UCSC.hg38<span class="op">::</span>BSgenome.Hsapiens.UCSC.hg38  )</a></code></pre></div>
<!-- 1. Install conda for python 2.7 -->
<!-- 2. Create a new conda environment: `conda create --name azimuthenv python=2.7` -->
<!-- 3. Activate conda environment:     `conda activate azimuthenv` -->
<!-- 4. Install module azimuth:         `pip install azimuth` -->
<!-- 5. Install scikit-learn:           `pip install scikit-learn==0.17.1` -->
</div>
<div id="define-targets" class="section level2">
<h2 class="hasAnchor">
<a href="#define-targets" class="anchor"></a>Define targets</h2>
<div id="from-a-bed-coordinate-file" class="section level3">
<h3 class="hasAnchor">
<a href="#from-a-bed-coordinate-file" class="anchor"></a>From a BED coordinate file</h3>
<p><code>bed_to_granges</code> converts a (0-based) BED coordinate file into a (1-based) GRanges.<br>
An example is loading the 1974 binding sites of the transcription factor SRF:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(magrittr)</a>
<a class="sourceLine" id="cb5-2" title="2"><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(multicrispr)</a>
<a class="sourceLine" id="cb5-3" title="3">bedfile &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata/SRF.bed'</span>, <span class="dt">package =</span> <span class="st">'multicrispr'</span>)</a>
<a class="sourceLine" id="cb5-4" title="4">tfbs0 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/bed_to_granges.html">bed_to_granges</a></span>(bedfile, <span class="dt">genome =</span> <span class="st">'mm10'</span>)</a>
<a class="sourceLine" id="cb5-5" title="5"><span class="co">##  Read SRF.bed into GRanges</span></a>
<a class="sourceLine" id="cb5-6" title="6"><span class="co">##      1974 ranges on 21 chromosomes</span></a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
<div id="from-gene-identifiers" class="section level3">
<h3 class="hasAnchor">
<a href="#from-gene-identifiers" class="anchor"></a>From gene identifiers</h3>
<p><code>genes_to_granges</code> and <code>genefile_to_granges</code> convert entrez/ensembl gene identifiers into GRanges using <code>txdb</code> gene models. An example that loads of a small subset of the 1974 SRF binding sites:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(multicrispr)</a>
<a class="sourceLine" id="cb6-2" title="2">entrezfile &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata/SRF.entrez'</span>, <span class="dt">package =</span> <span class="st">'multicrispr'</span>)</a>
<a class="sourceLine" id="cb6-3" title="3">txdb &lt;-<span class="st"> </span>TxDb.Mmusculus.UCSC.mm10.knownGene<span class="op">::</span>TxDb.Mmusculus.UCSC.mm10.knownGene</a>
<a class="sourceLine" id="cb6-4" title="4">sometfbs0 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/genes_to_granges.html">genefile_to_granges</a></span>(entrezfile, txdb, <span class="dt">complement =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb6-5" title="5"><span class="co">##      Convert 10 genes to 10 GRanges</span></a>
<a class="sourceLine" id="cb6-6" title="6"><span class="co">##      20 ranges after adding inverse strands</span></a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
</div>
<div id="from-a-coordinate-vector" class="section level3">
<h3 class="hasAnchor">
<a href="#from-a-coordinate-vector" class="anchor"></a>From a coordinate vector</h3>
<p><code>char_to_granges</code> uses a 1-based coordinate vector to specify the GRanges.<br>
It can be used to define four prime editing sites mentioned by Anzalone et al. (2019):</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1"><span class="co"># char_to_granges: Anzalone et al. (2019) prime editing targets</span></a>
<a class="sourceLine" id="cb7-2" title="2">    bsgenome &lt;-<span class="st"> </span>BSgenome.Hsapiens.UCSC.hg38<span class="op">::</span>BSgenome.Hsapiens.UCSC.hg38  </a>
<a class="sourceLine" id="cb7-3" title="3">    x &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dt">PRNP =</span> <span class="st">'chr20:4699600:+'</span>,             <span class="co"># snp</span></a>
<a class="sourceLine" id="cb7-4" title="4">           <span class="dt">HBB  =</span> <span class="st">'chr11:5227002:-'</span>,             <span class="co"># snp</span></a>
<a class="sourceLine" id="cb7-5" title="5">           <span class="dt">HEXA =</span> <span class="st">'chr15:72346580-72346583:-'</span>,   <span class="co"># del</span></a>
<a class="sourceLine" id="cb7-6" title="6">           <span class="dt">CFTR =</span> <span class="st">'chr7:117559593-117559595:+'</span>)  <span class="co"># ins</span></a>
<a class="sourceLine" id="cb7-7" title="7">    pe0 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/char_to_granges.html">char_to_granges</a></span>(x, bsgenome)</a>
<a class="sourceLine" id="cb7-8" title="8">    <span class="kw"><a href="../reference/plot_intervals.html">plot_intervals</a></span>(pe0)</a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
</div>
</div>
<div id="transform-targets" class="section level2">
<h2 class="hasAnchor">
<a href="#transform-targets" class="anchor"></a>Transform targets</h2>
<p>As a second step, extension or flanking may be required. The functions <code>left_flank</code>, <code>right_flank</code>, and <code>double_flank</code> flank target ranges, e.g. in order to target promoters or enhancers rather than TSS:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1"><span class="co"># Up flank</span></a>
<a class="sourceLine" id="cb8-2" title="2">targets0 &lt;-<span class="st"> </span>tfbs0</a>
<a class="sourceLine" id="cb8-3" title="3">targets &lt;-<span class="st"> </span><span class="kw"><a href="../reference/up_flank.html">up_flank</a></span>(  targets0, <span class="dv">-200</span>, <span class="dv">-1</span>, <span class="dt">plot =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1"></a>
<a class="sourceLine" id="cb9-2" title="2"><span class="co"># Down flank</span></a>
<a class="sourceLine" id="cb9-3" title="3">targets &lt;-<span class="st"> </span><span class="kw"><a href="../reference/up_flank.html">down_flank</a></span>( targets0, <span class="dv">1</span>, <span class="dv">200</span>, <span class="dt">plot =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-9-2.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1"></a>
<a class="sourceLine" id="cb10-2" title="2"><span class="co"># Double flank</span></a>
<a class="sourceLine" id="cb10-3" title="3">targets &lt;-<span class="st"> </span><span class="kw"><a href="../reference/double_flank.html">double_flank</a></span>(targets0, <span class="dv">-200</span>, <span class="dv">-1</span>, <span class="op">+</span><span class="dv">1</span>, <span class="op">+</span><span class="dv">200</span>, <span class="dt">plot =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb10-4" title="4"><span class="co">##      3948 flank ranges: 1974 up + 1974 down</span></a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-9-3.png" width="700"></p>
<p>The function <code>extend</code> expands target ranges in either (or both) direction(s), ensuring proper width to contain 23 base Cas9 sites.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1">targets &lt;-<span class="st"> </span><span class="kw"><a href="../reference/up_flank.html">extend</a></span>(targets0, <span class="dv">-22</span>, <span class="dv">22</span>, <span class="dt">plot =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div id="find-spacers" class="section level2">
<h2 class="hasAnchor">
<a href="#find-spacers" class="anchor"></a>Find spacers</h2>
<p>The third step consists of finding N<sub>20</sub>NGG Cas9 sites within the target ranges.<br>
First, we use <code>add_seq</code> to extract the relevant sequences using a BSgenome (R/BioC genome sequence object).<br>
Then we use <code>find_cas9s</code> to find N<sub>20</sub>NGG Cas9 sequences and ranges within targets. The associated plot shows that (nearly) all targets have Cas9 sites.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1">bsgenome &lt;-<span class="st"> </span>BSgenome.Mmusculus.UCSC.mm10<span class="op">::</span>BSgenome.Mmusculus.UCSC.mm10</a>
<a class="sourceLine" id="cb12-2" title="2">spacers &lt;-<span class="st"> </span><span class="kw"><a href="../reference/find_spacers.html">find_spacers</a></span>(targets, <span class="dt">bsgenome=</span>bsgenome)</a>
<a class="sourceLine" id="cb12-3" title="3"><span class="co">##      3948 ranges after adding inverse strands</span></a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
</div>
<div id="predict-targeting-efficiency" class="section level2">
<h2 class="hasAnchor">
<a href="#predict-targeting-efficiency" class="anchor"></a>Predict targeting efficiency</h2>
<p>Not all N<sub>20</sub>NGG gRNA sequences target equally well (even when matching sequence perfectly). For each position in the 23-bp gRNA sequence, the nucleotide present in current, previous and next position has an effect on targeting efficiency. Doench et al. (2014) were the first to systematically investigate the relationship between sequence and targeting efficiency, which they summarized in their 2014 prediction model. The function <code>score_cas9s</code> uses this model to compute the Doench 2014 targeting efficiency predictions:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1">spacers <span class="op">%&lt;&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/add_efficiency.html">add_efficiency</a></span>(bsgenome, <span class="st">'Doench2014'</span>)</a>
<a class="sourceLine" id="cb13-2" title="2"><span class="co">##  Score crispr spacers</span></a>
<a class="sourceLine" id="cb13-3" title="3"><span class="co">##      Add (4-23-3) contextseqs</span></a>
<a class="sourceLine" id="cb13-4" title="4"><span class="co">##      Score contextseqs with Doench2014</span></a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<p>The same authors later developed an updated model (Doench et al., 2016), which to-date remains the most comprehensive prediction model developed and the current community standard. The prediction model was implemented as a python module, which the authors freely share on github (github/MicrosoftResearch/azimuth). The installation of this python module is fairly straightforward:</p>
<ol style="list-style-type: decimal">
<li>Install conda for python 2.7</li>
<li>Create a new conda environment: <code>conda create --name azimuthenv python=2.7</code>
</li>
<li>Activate conda environment: <code>conda activate azimuthenv</code>
</li>
<li>Install module azimuth: <code>pip install azimuth</code>
</li>
<li>Install scikit-learn: <code>pip install scikit-learn==0.17.1</code>
</li>
</ol>
<p>Once azimuth is installed, <code>score_cas9s</code> can compute the Doench2016 score as well, under the hood calling the python module azimuth with a little help from the amazing R-to-python reticulate package.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1"><span class="co"># Install azimuth</span></a>
<a class="sourceLine" id="cb14-2" title="2"><span class="co"># Important: run R(Studio) with admin privileges for this to work</span></a>
<a class="sourceLine" id="cb14-3" title="3">  install_azimuth &lt;-<span class="st"> </span><span class="ot">FALSE</span></a>
<a class="sourceLine" id="cb14-4" title="4">  <span class="cf">if</span> (install_azimuth){</a>
<a class="sourceLine" id="cb14-5" title="5">    reticulate<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/reticulate/man/conda-tools.html">conda_create</a></span>(<span class="st">'azienv'</span>, <span class="st">'python=2.7'</span>)                         <span class="co"># Create condaenv 'azimuth'</span></a>
<a class="sourceLine" id="cb14-6" title="6">    reticulate<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/reticulate/man/conda-tools.html">conda_install</a></span>(<span class="st">'azienv'</span>, <span class="st">'azimuth'</span>, <span class="dt">pip =</span> <span class="ot">TRUE</span>)               <span class="co"># Install azimuth</span></a>
<a class="sourceLine" id="cb14-7" title="7">    reticulate<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/reticulate/man/conda-tools.html">conda_install</a></span>(<span class="st">'azienv'</span>, <span class="st">'scikit-learn==0.17.1'</span>, <span class="dt">pip =</span> <span class="ot">TRUE</span>)  <span class="co"># Install scikit-learn</span></a>
<a class="sourceLine" id="cb14-8" title="8">  }</a>
<a class="sourceLine" id="cb14-9" title="9"></a>
<a class="sourceLine" id="cb14-10" title="10"><span class="co"># Score Doench2016</span></a>
<a class="sourceLine" id="cb14-11" title="11">  <span class="cf">if</span> (reticulate<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/reticulate/man/py_module_available.html">py_module_available</a></span>(<span class="st">'azimuth'</span>)){</a>
<a class="sourceLine" id="cb14-12" title="12">      spacers <span class="op">%&lt;&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/add_efficiency.html">add_efficiency</a></span>(bsgenome, <span class="st">'Doench2016'</span>)</a>
<a class="sourceLine" id="cb14-13" title="13">  }</a>
<a class="sourceLine" id="cb14-14" title="14"><span class="co">##  Score crispr spacers</span></a>
<a class="sourceLine" id="cb14-15" title="15"><span class="co">##      Add (4-23-3) contextseqs</span></a>
<a class="sourceLine" id="cb14-16" title="16"><span class="co">##      Score contextseqs with Doench2016 (azimuth)</span></a>
<a class="sourceLine" id="cb14-17" title="17"><span class="co">##      Run Doench2016 1 times on 10000-seq chunks and concatenate (to preserve memory)</span></a>
<a class="sourceLine" id="cb14-18" title="18"><span class="co">## No model file specified, using V3_model_nopos</span></a>
<a class="sourceLine" id="cb14-19" title="19"><span class="co">##      Completed in 13 secs</span></a></code></pre></div>
<p><img src="designing_grna_libraries_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
</div>
<div id="prevent-offtarget-effects" class="section level2">
<h2 class="hasAnchor">
<a href="#prevent-offtarget-effects" class="anchor"></a>Prevent offtarget effects</h2>
<p>Finally, we can restrict the Cas9 sites further to only those that have no offtarget effects. For purposes of demonstration, let us restrict ourselves to Y-chromosome targets.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1"><span class="kw"><a href="../reference/index_genome.html">index_genome</a></span>(bsgenome)</a>
<a class="sourceLine" id="cb15-2" title="2">spacers <span class="op">%&lt;&gt;%</span><span class="st"> </span><span class="kw"><a href="../reference/filter_target_specific.html">add_specificity</a></span>(targets, bsgenome)</a>
<a class="sourceLine" id="cb15-3" title="3">spacers</a></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#background">Background</a><ul class="nav nav-pills nav-stacked">
<li><a href="#crisprcas9-and-prime-editing">Crispr/Cas9 and Prime Editing</a></li>
      <li><a href="#guide-rna-design">Guide RNA design</a></li>
      </ul>
</li>
      <li>
<a href="#multicrispr">Multicrispr</a><ul class="nav nav-pills nav-stacked">
<li><a href="#install">Install</a></li>
      <li><a href="#define-targets">Define targets</a></li>
      <li><a href="#transform-targets">Transform targets</a></li>
      <li><a href="#find-spacers">Find spacers</a></li>
      <li><a href="#predict-targeting-efficiency">Predict targeting efficiency</a></li>
      <li><a href="#prevent-offtarget-effects">Prevent offtarget effects</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Aditya Bhagwat.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
