<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using multicrispr for gRNA design • multicrispr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using multicrispr for gRNA design">
<meta property="og:description" content="multicrispr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">multicrispr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.23</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/genome_arithmetics.html">Genome arithmetics with multicrispr</a>
    </li>
    <li>
      <a href="../articles/grna_design.html">Using multicrispr for gRNA design</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using multicrispr for gRNA design</h1>
                        <h4 class="author">Aditya M Bhagwat</h4>
            
            <h4 class="date">2020-09-06</h4>
      
      
      <div class="hidden name"><code>grna_design.Rmd</code></div>

    </div>

    
    
<!-- ![](../inst/extdata/readme.png) -->
<div id="background" class="section level1">
<h1 class="hasAnchor">
<a href="#background" class="anchor"></a>Background</h1>
<div id="crisprcas9-and-prime-editing" class="section level2">
<h2 class="hasAnchor">
<a href="#crisprcas9-and-prime-editing" class="anchor"></a>Crispr/Cas9 and Prime Editing</h2>
<p><strong>Crispr/Cas9</strong> is a prokaryotic immune system turned into a genome engineering tool, with a Cas9/gRNA complex at the heart of its functioning. The <strong>gRNA</strong> (guide RNA) contains a 20 nucleotide (proto)<strong>spacer</strong> which guides the complex to a genomic locus with identical sequence, but only when followed by an <em>NGG</em> <strong>PAM</strong> (protospacer adjacent motif (see Figure 1A). The <strong>Cas9</strong> enzyme performs some effector action at that locus: <em>wildtype Cas9</em> cuts both strands after spacer nucleotide 17, <em>Cas9Nickase</em> cuts only a single strand (two variants exist, cutting respectively each of the strands), while catalytically dead <em>dCas9</em>, fused to some effector domain (KRAB, VPR, APEX) performs some alternative action (e.g. KRAB represses, VPR activates, APEX biotinylates).</p>
<p><strong>Prime Editing</strong> (Anzalone et al., 2019) extends this further, by coupling Cas9 to a Reverse Transcriptase (RT). As shown in Figure 1, the guide RNA is extended with a 3’ extension that contains primer binding site (binding a complementary DNA region that acts as a primer for reverse transcription) and reverse transcription template (the template used for reverse transcription). The spacer still acts as targeting agent, guiding the Cas9/RT complex to targeted genomic locus.</p>
<div class="figure">
<img src="../inst/extdata/pe.png" alt="gRNAs for Crispr/Cas9 (A) and Prime Editing" width="352"><p class="caption">
gRNAs for Crispr/Cas9 (A) and Prime Editing
</p>
</div>
</div>
<div id="guide-rna-design" class="section level2">
<h2 class="hasAnchor">
<a href="#guide-rna-design" class="anchor"></a>Guide RNA design</h2>
<p><strong>Guide RNA design</strong> involves finding good guide RNAs to hit the targets of interest. Plasmids with guide RNA scaffolds, ready for cloning desired spacers into, can be readily ordered from AddGene. The actual task, therefore, boils down to finding good spacers for the targets of interest. For prime editing, it additionally involves finding appropriate primer binding site and reverse transcription template to edit the target site of interest.</p>
<p>A <strong>good spacer</strong> needs to fulfill two requirements.</p>
<ol style="list-style-type: decimal">
<li>
<p><strong>Minimal off-target</strong> (mis)matches, so that only intended targets are hit.</p>
<ul>
<li>Crispr gRNAs can hit exact (with identical and alternate NGG pam), as well as (up-to 2) mismatch offtargets.</li>
<li>Prime Editing is much more specific, hitting only exact offtargets.</li>
</ul>
</li>
<li><p><strong>Maximal on-target</strong> efficiency. Over the years, several sequence-to-efficiency prediction models have been developed, of which the Doench 2016 score has currently become the community standard. The score is not perfect, and examples can be found where the prediction and the actual outcome differ. Yet, to-date it is the best heuristic to predict on-target efficiency, worth making use of.</p></li>
</ol>
</div>
</div>
<div id="multicrispr" class="section level1">
<h1 class="hasAnchor">
<a href="#multicrispr" class="anchor"></a>Multicrispr</h1>
<p>Multicrispr was developed to make guide RNA design easier. As shown below, it contains functions to <strong>define</strong> and <strong>transform</strong> targets, <strong>find spacers</strong>, compute/add <strong>offtarget counts</strong> and <strong>efficiency scores</strong> (Doench 2016), and finally return all of this as a <strong>GRanges</strong> object.</p>
<p><img src="../inst/extdata/readme_portrait.png" width="80%"></p>
<div id="install" class="section level2">
<h2 class="hasAnchor">
<a href="#install" class="anchor"></a>Install</h2>
<p>Installing <strong>multicrispr</strong> is simple:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="co"># From BioC</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)
<span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="kw">version</span><span class="kw">=</span><span class="st">'devel'</span>)
<span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"multicrispr"</span>)
<span class="co"># From gitlab: </span>
<span class="no">url</span> <span class="kw">&lt;-</span> <span class="st">'https://gitlab.gwdg.de/loosolab/software/multicrispr.git'</span>
<span class="kw pkg">remotes</span><span class="kw ns">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_git.html">install_git</a></span>(<span class="no">url</span>, <span class="kw">repos</span> <span class="kw">=</span> <span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/repositories.html">repositories</a></span>())</pre></body></html></div>
<p>Doench et al. (2016) ’s python package <strong>azimuth</strong> for on-target efficiency prediction using their method can be easily installed and activated using reticulate:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="co"># Install once</span>
  <span class="co"># reticulate::conda_create('azienv', 'python=2.7')</span>
  <span class="co"># reticulate::conda_install('azienv', 'azimuth', pip = TRUE)</span>
  <span class="co"># reticulate::conda_install('azienv', 'scikit-learn==0.17.1', pip = TRUE)</span>
<span class="co"># Then activate</span>
  <span class="kw pkg">reticulate</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/reticulate/man/use_python.html">use_condaenv</a></span>(<span class="st">'azienv'</span>)</pre></body></html></div>
<p>Bowtie-indexed genomes for quick offtarget analysis can be installed using <code>index_genome</code>. For the two genomes used in the examples, mm10 and hg38, the functions downloads pre-build indexes from our data server, allowing a quick exploration (set <code>download=FALSE</code> to build index anew):</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="../reference/index_genome.html">index_genome</a></span>(<span class="kw pkg">BSgenome.Mmusculus.UCSC.mm10</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/BSgenome.Mmusculus.UCSC.mm10/man/package.html">BSgenome.Mmusculus.UCSC.mm10</a></span>)
<span class="fu"><a href="../reference/index_genome.html">index_genome</a></span>(<span class="kw pkg">BSgenome.Hsapiens.UCSC.hg38</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg38/man/package.html">BSgenome.Hsapiens.UCSC.hg38</a></span>  )</pre></body></html></div>
<!-- 1. Install conda for python 2.7 -->
<!-- 2. Create a new conda environment: `conda create --name azimuthenv python=2.7` -->
<!-- 3. Activate conda environment:     `conda activate azimuthenv` -->
<!-- 4. Install module azimuth:         `pip install azimuth` -->
<!-- 5. Install scikit-learn:           `pip install scikit-learn==0.17.1` -->
</div>
<div id="define-targets" class="section level2">
<h2 class="hasAnchor">
<a href="#define-targets" class="anchor"></a>Define targets</h2>
<p><img src="../inst/extdata/01_define_targets.png" width="70%"></p>
<p><code>bed_to_granges</code> converts a (0-based) BED coordinate file into a (1-based) GRanges.<br>
An example is loading the 1974 binding sites of the transcription factor SRF:</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="no">magrittr</span>)
<span class="co">## Lade nötiges Paket: magrittr</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="no">multicrispr</span>)
<span class="co">## Lade nötiges Paket: multicrispr</span>
<span class="no">bedfile</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata/SRF.bed'</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">'multicrispr'</span>)
<span class="no">tfbs0</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/bed_to_granges.html">bed_to_granges</a></span>(<span class="no">bedfile</span>, <span class="kw">genome</span> <span class="kw">=</span> <span class="st">'mm10'</span>)
<span class="co">##  Read SRF.bed into GRanges</span>
<span class="co">## </span>
<span class="co">## Attache Paket: 'BiocGenerics'</span>
<span class="co">## The following objects are masked from 'package:parallel':</span>
<span class="co">## </span>
<span class="co">##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span>
<span class="co">##     clusterExport, clusterMap, parApply, parCapply, parLapply,</span>
<span class="co">##     parLapplyLB, parRapply, parSapply, parSapplyLB</span>
<span class="co">## The following objects are masked from 'package:stats':</span>
<span class="co">## </span>
<span class="co">##     IQR, mad, sd, var, xtabs</span>
<span class="co">## The following objects are masked from 'package:base':</span>
<span class="co">## </span>
<span class="co">##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,</span>
<span class="co">##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span>
<span class="co">##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,</span>
<span class="co">##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span>
<span class="co">##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,</span>
<span class="co">##     union, unique, unsplit, which, which.max, which.min</span>
<span class="co">## </span>
<span class="co">## Attache Paket: 'S4Vectors'</span>
<span class="co">## The following object is masked from 'package:base':</span>
<span class="co">## </span>
<span class="co">##     expand.grid</span>
<span class="co">## </span>
<span class="co">## Attache Paket: 'IRanges'</span>
<span class="co">## The following object is masked from 'package:grDevices':</span>
<span class="co">## </span>
<span class="co">##     windows</span>
<span class="co">## </span>
<span class="co">## Attache Paket: 'Biostrings'</span>
<span class="co">## The following object is masked from 'package:base':</span>
<span class="co">## </span>
<span class="co">##     strsplit</span>
<span class="co">##      1974 ranges on 21 chromosomes</span></pre></body></html></div>
<p><img src="grna_design_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p><code>genes_to_granges</code> and <code>genefile_to_granges</code> convert entrez/ensembl gene identifiers into GRanges using <code>txdb</code> gene models. An example that loads of a small subset of the 1974 SRF binding sites:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="no">multicrispr</span>)
<span class="no">entrezfile</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">'extdata/SRF.entrez'</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">'multicrispr'</span>)
<span class="no">txdb</span> <span class="kw">&lt;-</span> <span class="kw pkg">TxDb.Mmusculus.UCSC.mm10.knownGene</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/TxDb.Mmusculus.UCSC.mm10.knownGene/man/package.html">TxDb.Mmusculus.UCSC.mm10.knownGene</a></span>
<span class="no">sometfbs0</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/genes_to_granges.html">genefile_to_granges</a></span>(<span class="no">entrezfile</span>, <span class="no">txdb</span>, <span class="kw">complement</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="co">##   66 genes were dropped because they have exons located on both strands</span>
<span class="co">##   of the same reference sequence or on more than one reference sequence,</span>
<span class="co">##   so cannot be represented by a single genomic range.</span>
<span class="co">##   Use 'single.strand.genes.only=FALSE' to get all the genes in a</span>
<span class="co">##   GRangesList object, or use suppressMessages() to suppress this message.</span>
<span class="co">##      Convert 10 genes to 10 GRanges</span>
<span class="co">##  Retain 20 unique target ranges after adding inverse strands</span></pre></body></html></div>
<p><img src="grna_design_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p><code>char_to_granges</code> uses a 1-based coordinate vector to specify the GRanges.<br>
It can be used to define four prime editing sites mentioned by Anzalone et al. (2019):</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="co"># char_to_granges: Anzalone et al. (2019) prime editing targets</span>
<span class="no">bsgenome</span> <span class="kw">&lt;-</span> <span class="kw pkg">BSgenome.Hsapiens.UCSC.hg38</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg38/man/package.html">BSgenome.Hsapiens.UCSC.hg38</a></span>
<span class="no">pe0</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/char_to_granges.html">char_to_granges</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw">HBB</span>  <span class="kw">=</span> <span class="st">'chr11:5227002:-'</span>), <span class="no">bsgenome</span>)
<span class="fu"><a href="../reference/plot_intervals.html">plot_intervals</a></span>(<span class="no">pe0</span>)</pre></body></html></div>
<p><img src="grna_design_files/figure-html/unnamed-chunk-10-1.png" width="336"></p>
</div>
<div id="transform-targets" class="section level2">
<h2 class="hasAnchor">
<a href="#transform-targets" class="anchor"></a>Transform targets</h2>
<p><img src="../inst/extdata/02_transform.png" width="70%"></p>
<p>As a second step, the original targets may require transformation. The functions <code>extend</code>, <code>up_flank</code> (upstream flanks), <code>down_flank</code> (downstream flanks), and <code>double_flank</code> (double flanks) can be used to update the target ranges as required.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="co"># Extend</span>
<span class="no">targets</span> <span class="kw">&lt;-</span> <span class="no">pe0</span>
<span class="fu"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="fu"><a href="../reference/up_flank.html">extend</a></span>(<span class="no">targets</span>, -<span class="fl">22</span>, <span class="fl">22</span>, <span class="kw">plot</span> <span class="kw">=</span> <span class="fl">TRUE</span>))</pre></body></html></div>
<p><img src="grna_design_files/figure-html/unnamed-chunk-12-1.png" width="65%"></p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="co"># Up flank</span>
<span class="fu"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="fu"><a href="../reference/up_flank.html">up_flank</a></span>(  <span class="no">targets</span>, -<span class="fl">200</span>, -<span class="fl">1</span>, <span class="kw">plot</span> <span class="kw">=</span> <span class="fl">TRUE</span>))</pre></body></html></div>
<p><img src="grna_design_files/figure-html/unnamed-chunk-12-2.png" width="65%"></p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="co"># Down flank</span>
<span class="fu"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="fu"><a href="../reference/up_flank.html">down_flank</a></span>( <span class="no">targets</span>, <span class="fl">1</span>, <span class="fl">200</span>, <span class="kw">plot</span> <span class="kw">=</span> <span class="fl">TRUE</span>))</pre></body></html></div>
<p><img src="grna_design_files/figure-html/unnamed-chunk-12-3.png" width="65%"></p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="co"># Double flank</span>
<span class="fu"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="fu"><a href="../reference/double_flank.html">double_flank</a></span>(<span class="no">targets</span>, -<span class="fl">200</span>, -<span class="fl">1</span>, +<span class="fl">1</span>, +<span class="fl">200</span>, <span class="kw">plot</span> <span class="kw">=</span> <span class="fl">TRUE</span>))
<span class="co">##      2 flank ranges: 1 up + 1 down</span></pre></body></html></div>
<p><img src="grna_design_files/figure-html/unnamed-chunk-12-4.png" width="65%"></p>
</div>
<div id="find-spacers-count-off-targets-score-on-targets-find-nickspacers" class="section level2">
<h2 class="hasAnchor">
<a href="#find-spacers-count-off-targets-score-on-targets-find-nickspacers" class="anchor"></a>Find spacers, count off-targets, score on-targets, find nickspacers</h2>
<p><img src="../inst/extdata/03_find.png" width="70%"></p>
<p>The next step is to find N20 spacers followed by NGG pam sites. This can be done with <code>find_spacers</code>, after extending target range(s) to prime editing neighbourhood(s). This function also counts off-targets and scores on-targets.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">bsgenome</span>   <span class="kw">&lt;-</span> <span class="kw pkg">BSgenome.Hsapiens.UCSC.hg38</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg38/man/package.html">BSgenome.Hsapiens.UCSC.hg38</a></span>
<span class="no">neighbourhood</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/extend_for_pe.html">extend_for_pe</a></span>(<span class="no">targets</span>, <span class="no">bsgenome</span>)
<span class="no">spacers</span>    <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/find_spacers.html">find_spacers</a></span>(<span class="no">neighbourhood</span>, <span class="no">bsgenome</span>, <span class="kw">complement</span> <span class="kw">=</span> <span class="fl">FALSE</span>, <span class="kw">mismatches</span> <span class="kw">=</span> <span class="fl">0</span>)
<span class="co">## Find spacers in 2 targets</span>
<span class="co">##  Found 2 spacers</span>
<span class="co">##  Count offtargets with max 0 mismatches using bowtie</span>
<span class="co">##      Count genome (mis)matches</span>
<span class="co">##          Expand iupac ambiguities in pam</span>
<span class="co">##          Write reads to ~/multicrisprout/spacers.fa</span>
<span class="co">##          Map reads: ~/multicrisprout/spacers/spacers_to_BSgenome.Hsapiens.UCSC.hg38.txt</span>
<span class="co">##          Load results</span>
<span class="co">##              Read 12 hits with max 1 mismatch(es)</span>
<span class="co">##              Retain 3 after removing NGG 'N' mismatches (avoid double counting expanded pams)</span>
<span class="co">##              Retain 3 after removing NGG 'G1' mismatches(except for NGG -&gt; NAG, which is allowed)</span>
<span class="co">##              Retain 3 after removing NGG 'G2' mismatches</span>
<span class="co">##              Retain 3 after removing more than 0 mismatches</span>
<span class="co">##          Count matches</span>
<span class="co">##      Count off-targets</span>
<span class="co">##             1 are off0 free</span>
<span class="co">##  Score ontargets</span>
<span class="co">##      Add (4-23-3) contextseqs</span>
<span class="co">##      Score contextseqs with Doench2014</span></pre></body></html></div>
<p><img src="grna_design_files/figure-html/unnamed-chunk-14-1.png" width="288"></p>
<p>For prime editing, however, it is better to use <code>find_primespacers</code>, which performs target to neighbourhood extension implicitly, and also finds 3’ extensions and nickspacers (Fig 1) to returned spacer GRanges.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="no">spacers</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/find_primespacers.html">find_primespacers</a></span>(<span class="no">targets</span>, <span class="kw">bsgenome</span><span class="kw">=</span><span class="no">bsgenome</span>)
<span class="co">## Find primespacers for 1 targets</span>
<span class="co">##  Found 2 primespacers</span>
<span class="co">##  Count offtargets with max 0 mismatches using bowtie</span>
<span class="co">##      Count genome (mis)matches</span>
<span class="co">##          Expand iupac ambiguities in pam</span>
<span class="co">##          Write reads to ~/multicrisprout/spacers.fa</span>
<span class="co">##          Map reads: ~/multicrisprout/spacers/spacers_to_BSgenome.Hsapiens.UCSC.hg38.txt</span>
<span class="co">##          Load results</span>
<span class="co">##              Read 12 hits with max 1 mismatch(es)</span>
<span class="co">##              Retain 3 after removing NGG 'N' mismatches (avoid double counting expanded pams)</span>
<span class="co">##              Retain 3 after removing NGG 'G1' mismatches(except for NGG -&gt; NAG, which is allowed)</span>
<span class="co">##              Retain 3 after removing NGG 'G2' mismatches</span>
<span class="co">##              Retain 3 after removing more than 0 mismatches</span>
<span class="co">##          Count matches</span>
<span class="co">##      Count off-targets</span>
<span class="co">##             1 are off0 free</span>
<span class="co">##  Score ontargets</span>
<span class="co">##      Add (4-23-3) contextseqs</span>
<span class="co">##      Score contextseqs with Doench2014</span>
<span class="co">## Find nickspacers</span>
<span class="co">## Find spacers in 2 targets</span>
<span class="co">##  Found 2 spacers</span>
<span class="co">##  Count offtargets with max 2 mismatches using bowtie</span>
<span class="co">##      Count genome (mis)matches</span>
<span class="co">##          Expand iupac ambiguities in pam</span>
<span class="co">##          Write reads to ~/multicrisprout/spacers.fa</span>
<span class="co">##          Map reads: ~/multicrisprout/spacers/spacers_to_BSgenome.Hsapiens.UCSC.hg38.txt</span>
<span class="co">##          Load results</span>
<span class="co">##              Read 16 hits with max 2 mismatch(es)</span>
<span class="co">##              Retain 4 after removing NGG 'N' mismatches (avoid double counting expanded pams)</span>
<span class="co">##              Retain 4 after removing NGG 'G1' mismatches(except for NGG -&gt; NAG, which is allowed)</span>
<span class="co">##              Retain 4 after removing NGG 'G2' mismatches</span>
<span class="co">##              Retain 4 after removing more than 2 mismatches</span>
<span class="co">##          Count matches</span>
<span class="co">##      Count off-targets</span>
<span class="co">##             0 are off0 free</span>
<span class="co">##             2 are off1 free</span>
<span class="co">##             2 are off2 free</span>
<span class="co">##  Filter for best offtarget counts</span>
<span class="co">##      Retain 2 with best off0 counts per pename</span>
<span class="co">##      Retain 2 with best off1 counts per pename</span>
<span class="co">##      Retain 2 with best off2 counts per pename</span>
<span class="co">##  Score ontargets</span>
<span class="co">##      Add (4-23-3) contextseqs</span>
<span class="co">##      Score contextseqs with Doench2014</span>
<span class="co">## Warning: Removed 4 rows containing missing values (geom_point).</span>

<span class="co">## Warning: Removed 4 rows containing missing values (geom_point).</span></pre></body></html></div>
<p><img src="grna_design_files/figure-html/unnamed-chunk-15-1.png" width="316.8"></p>
<p>This finds prime editing sites that follow Anzalone et al. (2019)’s recommended default settings: a 13-nucleotide primer and a 16-nucleotide reverse transcription template. These settings can be easily altered (e.g. Anzalone et al. report success with reverse transcription templates of up to 48 nucleotides) by.</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="no">spacers</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/find_primespacers.html">find_primespacers</a></span>(<span class="no">targets</span>, <span class="kw">bsgenome</span><span class="kw">=</span><span class="no">bsgenome</span>, <span class="kw">nrt</span> <span class="kw">=</span> <span class="fl">48</span>)
<span class="co">## Find primespacers for 1 targets</span>
<span class="co">##  Found 6 primespacers</span>
<span class="co">##  Count offtargets with max 0 mismatches using bowtie</span>
<span class="co">##      Count genome (mis)matches</span>
<span class="co">##          Expand iupac ambiguities in pam</span>
<span class="co">##          Write reads to ~/multicrisprout/spacers.fa</span>
<span class="co">##          Map reads: ~/multicrisprout/spacers/spacers_to_BSgenome.Hsapiens.UCSC.hg38.txt</span>
<span class="co">##          Load results</span>
<span class="co">##              Read 36 hits with max 1 mismatch(es)</span>
<span class="co">##              Retain 12 after removing NGG 'N' mismatches (avoid double counting expanded pams)</span>
<span class="co">##              Retain 12 after removing NGG 'G1' mismatches(except for NGG -&gt; NAG, which is allowed)</span>
<span class="co">##              Retain 12 after removing NGG 'G2' mismatches</span>
<span class="co">##              Retain 8 after removing more than 0 mismatches</span>
<span class="co">##          Count matches</span>
<span class="co">##      Count off-targets</span>
<span class="co">##             4 are off0 free</span>
<span class="co">##  Score ontargets</span>
<span class="co">##      Add (4-23-3) contextseqs</span>
<span class="co">##      Score contextseqs with Doench2014</span>
<span class="co">## Find nickspacers</span>
<span class="co">## Find spacers in 6 targets</span>
<span class="co">##  Found 7 spacers</span>
<span class="co">##  Count offtargets with max 2 mismatches using bowtie</span>
<span class="co">##      Count genome (mis)matches</span>
<span class="co">##          Expand iupac ambiguities in pam</span>
<span class="co">##          Write reads to ~/multicrisprout/spacers.fa</span>
<span class="co">##          Map reads: ~/multicrisprout/spacers/spacers_to_BSgenome.Hsapiens.UCSC.hg38.txt</span>
<span class="co">##          Load results</span>
<span class="co">##              Read 51 hits with max 2 mismatch(es)</span>
<span class="co">##              Retain 15 after removing NGG 'N' mismatches (avoid double counting expanded pams)</span>
<span class="co">##              Retain 15 after removing NGG 'G1' mismatches(except for NGG -&gt; NAG, which is allowed)</span>
<span class="co">##              Retain 13 after removing NGG 'G2' mismatches</span>
<span class="co">##              Retain 13 after removing more than 2 mismatches</span>
<span class="co">##          Count matches</span>
<span class="co">##      Count off-targets</span>
<span class="co">##             2 are off0 free</span>
<span class="co">##             5 are off1 free</span>
<span class="co">##             6 are off2 free</span>
<span class="co">##  Filter for best offtarget counts</span>
<span class="co">##      Retain 6 with best off0 counts per pename</span>
<span class="co">##      Retain 5 with best off1 counts per pename</span>
<span class="co">##      Retain 5 with best off2 counts per pename</span>
<span class="co">##  Score ontargets</span>
<span class="co">##      Add (4-23-3) contextseqs</span>
<span class="co">##      Score contextseqs with Doench2014</span>
<span class="co">## Warning: Removed 11 rows containing missing values (geom_point).</span>

<span class="co">## Warning: Removed 11 rows containing missing values (geom_point).</span></pre></body></html></div>
<p><img src="grna_design_files/figure-html/unnamed-chunk-16-1.png" width="345.6"></p>
</div>
<div id="return-spacer-granges" class="section level2">
<h2 class="hasAnchor">
<a href="#return-spacer-granges" class="anchor"></a>Return spacer GRanges</h2>
<div class="sourceCode" id="cb14"><html><body><pre class="r">  <span class="no">spacers</span>
<span class="co">## GRanges object with 6 ranges and 21 metadata columns:</span>
<span class="co">##         seqnames          ranges strand |  targetname targetstart targetend</span>
<span class="co">##            &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;   &lt;integer&gt; &lt;integer&gt;</span>
<span class="co">##   HBB_1    chr11 5226958-5226977      + |         HBB     5227002   5227002</span>
<span class="co">##   HBB_2    chr11 5226959-5226978      + |         HBB     5227002   5227002</span>
<span class="co">##   HBB_3    chr11 5226968-5226987      + |         HBB     5227002   5227002</span>
<span class="co">##   HBB_4    chr11 5227003-5227022      - |         HBB     5227002   5227002</span>
<span class="co">##   HBB_5    chr11 5226984-5227003      + |         HBB     5227002   5227002</span>
<span class="co">##   HBB_6    chr11 5227021-5227040      - |         HBB     5227002   5227002</span>
<span class="co">##          crisprname         crisprspacer   crisprpam  crisprprimer</span>
<span class="co">##         &lt;character&gt;          &lt;character&gt; &lt;character&gt;   &lt;character&gt;</span>
<span class="co">##   HBB_1       HBB_1 CCACGTTCACCTTGCCCCAC         AGG GTTCACCTTGCCC</span>
<span class="co">##   HBB_2       HBB_2 CACGTTCACCTTGCCCCACA         GGG TTCACCTTGCCCC</span>
<span class="co">##   HBB_3       HBB_3 CTTGCCCCACAGGGCAGTAA         CGG CCCCACAGGGCAG</span>
<span class="co">##   HBB_4       HBB_4 CATGGTGCATCTGACTCCTG         AGG GTGCATCTGACTC</span>
<span class="co">##   HBB_5       HBB_5 GTAACGGCAGACTTCTCCTC         AGG CGGCAGACTTCTC</span>
<span class="co">##   HBB_6       HBB_6 GCAACCTCAAACAGACACCA         TGG CCTCAAACAGACA</span>
<span class="co">##                                         crisprtranscript</span>
<span class="co">##                                              &lt;character&gt;</span>
<span class="co">##   HBB_1 CACAGGGCAGTAACGGCAGACTTCTCCTCAGGAGTCAGATGCACCATG</span>
<span class="co">##   HBB_2 ACAGGGCAGTAACGGCAGACTTCTCCTCAGGAGTCAGATGCACCATGG</span>
<span class="co">##   HBB_3 TAACGGCAGACTTCTCCTCAGGAGTCAGATGCACCATGGTGTCTGTTT</span>
<span class="co">##   HBB_4 CTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAACGTGG</span>
<span class="co">##   HBB_5 CTCAGGAGTCAGATGCACCATGGTGTCTGTTTGAGGTTGCTAGTGAAC</span>
<span class="co">##   HBB_6 CCATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGT</span>
<span class="co">##                                                       crisprextension</span>
<span class="co">##                                                           &lt;character&gt;</span>
<span class="co">##   HBB_1 CATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAAC</span>
<span class="co">##   HBB_2 CCATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGGCAAGGTGAA</span>
<span class="co">##   HBB_3 AAACAGACACCATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCGTTACTGCCCTGTGGGG</span>
<span class="co">##   HBB_4 CCACGTTCACCTTGCCCCACAGGGCAGTAACGGCAGACTTCTCCTCAGGAGTCAGATGCAC</span>
<span class="co">##   HBB_5 GTTCACTAGCAACCTCAAACAGACACCATGGTGCATCTGACTCCTGAGGAGAAGTCTGCCG</span>
<span class="co">##   HBB_6 ACAGGGCAGTAACGGCAGACTTCTCCTCAGGAGTCAGATGCACCATGGTGTCTGTTTGAGG</span>
<span class="co">##                  crisprextrange       off      off0 Doench2014</span>
<span class="co">##                     &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;</span>
<span class="co">##   HBB_1 chr11:5226962-5227022:-         0         0  0.0244198</span>
<span class="co">##   HBB_2 chr11:5226963-5227023:-         0         0  0.3229083</span>
<span class="co">##   HBB_3 chr11:5226972-5227032:-         0         0  0.3044047</span>
<span class="co">##   HBB_4 chr11:5226958-5227018:+         1         1  0.1311142</span>
<span class="co">##   HBB_5 chr11:5226988-5227048:-         0         0  0.4139806</span>
<span class="co">##   HBB_6 chr11:5226976-5227036:+         1         1  0.7809699</span>
<span class="co">##                                               nickrange</span>
<span class="co">##                                             &lt;character&gt;</span>
<span class="co">##   HBB_1                         chr11:5227021-5227040:-</span>
<span class="co">##   HBB_2                         chr11:5227021-5227040:-</span>
<span class="co">##   HBB_3                                            &lt;NA&gt;</span>
<span class="co">##   HBB_4 chr11:5226916-5226935:+;chr11:5226917-5226936:+</span>
<span class="co">##   HBB_5                                            &lt;NA&gt;</span>
<span class="co">##   HBB_6                         chr11:5226959-5226978:+</span>
<span class="co">##                                        nickspacer     nickpam     nickoff</span>
<span class="co">##                                       &lt;character&gt; &lt;character&gt; &lt;character&gt;</span>
<span class="co">##   HBB_1                      GCAACCTCAAACAGACACCA         TGG           1</span>
<span class="co">##   HBB_2                      GCAACCTCAAACAGACACCA         TGG           1</span>
<span class="co">##   HBB_3                                      &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="co">##   HBB_4 ACCTTGATACCAACCTGCCC;CCTTGATACCAACCTGCCCA     AGG;GGG         1;1</span>
<span class="co">##   HBB_5                                      &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;</span>
<span class="co">##   HBB_6                      CACGTTCACCTTGCCCCACA         GGG           1</span>
<span class="co">##            nickoff0    nickoff1    nickoff2 nickDoench2014</span>
<span class="co">##         &lt;character&gt; &lt;character&gt; &lt;character&gt;    &lt;character&gt;</span>
<span class="co">##   HBB_1           1           0           0           0.78</span>
<span class="co">##   HBB_2           1           0           0           0.78</span>
<span class="co">##   HBB_3        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;           &lt;NA&gt;</span>
<span class="co">##   HBB_4         1;1         0;0         0;0      0.06;0.42</span>
<span class="co">##   HBB_5        &lt;NA&gt;        &lt;NA&gt;        &lt;NA&gt;           &lt;NA&gt;</span>
<span class="co">##   HBB_6           0           1           0           0.32</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 595 sequences (1 circular) from hg38 genome</span></pre></body></html></div>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>Anzalone, A.V., Randolph, P.B., Davis, J.R. et al. Search-and-replace genome editing without double-strand breaks or donor DNA. Nature 576, 149–157 (2019). <a href="https://doi.org/10.1038/s41586-019-1711-4" class="uri">https://doi.org/10.1038/s41586-019-1711-4</a></p>
<p>Doench et al. (2016). Optimized sgRNA design to maximize activity and minimize off-target effects of CRISPR-Cas9. Nature Biotechnology, 34(7), <a href="doi:10.1038/nbt.3437" class="uri">doi:10.1038/nbt.3437</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Aditya Bhagwat.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
