<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Prime Editing • multicrispr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Prime Editing">
<meta property="og:description" content="multicrispr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">multicrispr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.24</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/crispr_grna_design.html">Crispr/Cas9 gRNA design</a>
    </li>
    <li>
      <a href="../articles/genome_arithmetics.html">Genome arithmetics</a>
    </li>
    <li>
      <a href="../articles/prime_editing.html">Prime Editing</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Prime Editing</h1>
                        <h4 class="author">Aditya M Bhagwat</h4>
            
            <h4 class="date">2020-09-06</h4>
      
      
      <div class="hidden name"><code>prime_editing.Rmd</code></div>

    </div>

    
    
<!-- ![](../inst/extdata/readme.png) -->
<div id="background" class="section level1">
<h1 class="hasAnchor">
<a href="#background" class="anchor"></a>Background</h1>
<p><strong>Crispr/Cas9</strong> is a prokaryotic immune system turned into a genome engineering tool, with a Cas9/gRNA complex at the heart of its functioning. The <strong>gRNA</strong> (guide RNA) contains a 20 nucleotide (proto)<strong>spacer</strong> which guides the complex to a genomic locus with identical sequence, but only when followed by an <em>NGG</em> <strong>PAM</strong> (protospacer adjacent motif (see Figure 1A). The <strong>Cas9</strong> enzyme performs some effector action at that locus: <em>wildtype Cas9</em> cuts both strands after spacer nucleotide 17, <em>Cas9Nickase</em> cuts only a single strand (two variants exist, cutting respectively each of the strands), while catalytically dead <em>dCas9</em>, fused to some effector domain (KRAB, VPR, APEX) performs some alternative action (e.g. KRAB represses, VPR activates, APEX biotinylates).</p>
<p><strong>Prime Editing</strong> (Anzalone et al., 2019) extends this further, by coupling Cas9 to a Reverse Transcriptase (RT). As shown in Figure 1, the guide RNA is extended with a 3’ extension that contains primer binding site (binding a complementary DNA region that acts as a primer for reverse transcription) and reverse transcription template (the template used for reverse transcription). The spacer still acts as targeting agent, guiding the Cas9/RT complex to targeted genomic locus.</p>
<div class="figure">
<img src="../inst/extdata/pe.png" alt="gRNAs for Crispr/Cas9 (A) and Prime Editing" width="352"><p class="caption">
gRNAs for Crispr/Cas9 (A) and Prime Editing
</p>
</div>
<p><strong>Guide RNA design</strong> involves finding good guide RNAs to hit the targets of interest. Plasmids with guide RNA scaffolds, ready for cloning desired spacers into, can be readily ordered from AddGene. The actual task, therefore, boils down to finding good spacers for the targets of interest. For prime editing, it additionally involves finding appropriate primer binding site and reverse transcription template to edit the target site of interest. In general, a good spacer needs to fulfill two requirements:</p>
<ol style="list-style-type: decimal">
<li>
<p><strong>Minimal off-target</strong> (mis)matches, so that only intended targets are hit.</p>
<ul>
<li>Crispr gRNAs can hit exact (with identical and alternate NGG pam), as well as (up-to 3) mismatch offtargets.</li>
<li>Prime Editing is much more specific, hitting only exact offtargets.</li>
</ul>
</li>
<li><p><strong>Maximal on-target</strong> efficiency. Over the years, several sequence-to-efficiency prediction models have been developed, of which the Doench 2016 score has currently become the community standard. The score is not perfect, and examples can be found where the prediction and the actual outcome differ. Yet, to-date it is the best heuristic to predict on-target efficiency, worth making use of.</p></li>
</ol>
</div>
<div id="multicrispr" class="section level1">
<h1 class="hasAnchor">
<a href="#multicrispr" class="anchor"></a>Multicrispr</h1>
<p>Multicrispr was developed to make guide RNA design easier. As shown below, it contains functions to <strong>define</strong> and <strong>transform</strong> targets, <strong>find spacers</strong>, compute/add <strong>offtarget counts</strong> and <strong>efficiency scores</strong> (Doench 2016), and finally return all of this as a <strong>GRanges</strong> object.</p>

<p><img src="../inst/extdata/overview.png" width="100%"></p>
<div id="install" class="section level2">
<h2 class="hasAnchor">
<a href="#install" class="anchor"></a>Install</h2>
<p>Installing <strong>multicrispr</strong> is simple:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="co"># From BioC</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)
<span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="kw">version</span><span class="kw">=</span><span class="st">'devel'</span>)
<span class="kw pkg">BiocManager</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span>(<span class="st">"multicrispr"</span>)

<span class="co"># From gitlab: </span>
<span class="co">#url &lt;- 'https://gitlab.gwdg.de/loosolab/software/multicrispr.git'</span>
<span class="co">#remotes::install_git(url, repos = BiocManager::repositories())</span></pre></body></html></div>
<p>Doench et al. (2016) ’s python package <strong>azimuth</strong> for on-target efficiency prediction using their method can be easily installed and activated using reticulate:</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="co"># Install once</span>
  <span class="co"># reticulate::conda_create('azienv', 'python=2.7')</span>
  <span class="co"># reticulate::conda_install('azienv', 'azimuth', pip = TRUE)</span>
  <span class="co"># reticulate::conda_install('azienv', 'scikit-learn==0.17.1', pip = TRUE)</span>
<span class="co"># Then activate</span>
  <span class="kw pkg">reticulate</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/reticulate/man/use_python.html">use_condaenv</a></span>(<span class="st">'azienv'</span>)</pre></body></html></div>
<p>Bowtie-indexed genomes for quick offtarget analysis can be installed using <code>index_genome</code>. For the two genomes used in the examples, mm10 and hg38, the functions downloads pre-build indexes from our data server, allowing a quick exploration (set <code>download=FALSE</code> to build index anew):</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="../reference/index_genome.html">index_genome</a></span>(<span class="kw pkg">BSgenome.Mmusculus.UCSC.mm10</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/BSgenome.Mmusculus.UCSC.mm10/man/package.html">BSgenome.Mmusculus.UCSC.mm10</a></span>)
<span class="fu"><a href="../reference/index_genome.html">index_genome</a></span>(<span class="kw pkg">BSgenome.Hsapiens.UCSC.hg38</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg38/man/package.html">BSgenome.Hsapiens.UCSC.hg38</a></span>  )</pre></body></html></div>
<!-- 1. Install conda for python 2.7 -->
<!-- 2. Create a new conda environment: `conda create --name azimuthenv python=2.7` -->
<!-- 3. Activate conda environment:     `conda activate azimuthenv` -->
<!-- 4. Install module azimuth:         `pip install azimuth` -->
<!-- 5. Install scikit-learn:           `pip install scikit-learn==0.17.1` -->
</div>
<div id="define-targets" class="section level2">
<h2 class="hasAnchor">
<a href="#define-targets" class="anchor"></a>Define targets</h2>
<p><strong>bed_to_granges</strong> converts a (0-based) BED coordinate file into a (1-based) GRanges.<br><strong>genes_to_granges</strong> and <strong>genefile_to_granges</strong> convert entrez/ensembl gene identifiers into GRanges using <code>txdb</code> gene models. <strong>char_to_granges</strong> uses a 1-based coordinate vector to specify the GRanges. It can be used to prime edit the prion disease locus in the PRNP gene (Anzalone et al., 2019)</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="co"># char_to_granges: Anzalone et al. (2019) prime editing targets</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="no">multicrispr</span>)
<span class="co">## Lade nötiges Paket: multicrispr</span>
<span class="no">bsgenome</span> <span class="kw">&lt;-</span> <span class="kw pkg">BSgenome.Hsapiens.UCSC.hg38</span><span class="kw ns">::</span><span class="no"><a href="https://rdrr.io/pkg/BSgenome.Hsapiens.UCSC.hg38/man/package.html">BSgenome.Hsapiens.UCSC.hg38</a></span>
<span class="no">targets</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/char_to_granges.html">char_to_granges</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw">PRNP</span> <span class="kw">=</span> <span class="st">'chr20:4699600:+'</span>), <span class="no">bsgenome</span>)
<span class="fu"><a href="../reference/plot_intervals.html">plot_intervals</a></span>(<span class="no">targets</span>)</pre></body></html></div>
<p><img src="prime_editing_files/figure-html/unnamed-chunk-7-1.png" width="336"></p>
</div>
<div id="find-prime-editing-spacers-and-nicking-spacers" class="section level2">
<h2 class="hasAnchor">
<a href="#find-prime-editing-spacers-and-nicking-spacers" class="anchor"></a>Find prime editing spacers and nicking spacers</h2>
<p><strong>find_primespacers</strong> finds prime editing spacers, 3’ extensions and nicking spacers. For both, on- and off-target analysis is performed. Desired edits are always to be specified with respect to the defauylt (+) strand.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">spacers</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/find_primespacers.html">find_primespacers</a></span>(<span class="no">targets</span>, <span class="kw">bsgenome</span><span class="kw">=</span><span class="no">bsgenome</span>, <span class="kw">edits</span> <span class="kw">=</span> <span class="st">"T"</span>)</pre></body></html></div>
<p><img src="prime_editing_files/figure-html/unnamed-chunk-8-1.png" width="316.8"></p>
</div>
<div id="return-spacer-granges" class="section level2">
<h2 class="hasAnchor">
<a href="#return-spacer-granges" class="anchor"></a>Return spacer GRanges</h2>
<div class="sourceCode" id="cb6"><html><body><pre class="r">  <span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span>(<span class="fu"><a href="../reference/gr2dt.html">gr2dt</a></span>(<span class="no">spacers</span>), <span class="kw">vec.len</span><span class="kw">=</span><span class="fl">2</span>)
<span class="co">## Classes 'data.table' and 'data.frame':   4 obs. of  27 variables:</span>
<span class="co">##  $ seqnames        : Factor w/ 595 levels "chr1","chr2",..: 20 20 20 20</span>
<span class="co">##  $ start           : int  4699568 4699569 4699575 4699578</span>
<span class="co">##  $ end             : int  4699587 4699588 4699594 4699597</span>
<span class="co">##  $ width           : int  20 20 20 20</span>
<span class="co">##  $ strand          : Factor w/ 3 levels "+","-","*": 1 1 1 1</span>
<span class="co">##  $ targetname      : chr  "PRNP" "PRNP" ...</span>
<span class="co">##  $ targetstart     : int  4699600 4699600 4699600 4699600</span>
<span class="co">##  $ targetend       : int  4699600 4699600 4699600 4699600</span>
<span class="co">##  $ crisprname      : chr  "PRNP_1" "PRNP_2" ...</span>
<span class="co">##  $ crisprspacer    : chr  "AGCAGCTGGGGCAGTGGTGG" "GCAGCTGGGGCAGTGGTGGG" ...</span>
<span class="co">##  $ crisprpam       : chr  "GGG" "GGG" ...</span>
<span class="co">##  $ crisprprimer    : chr  "GCTGGGGCAGTGG" "CTGGGGCAGTGGT" ...</span>
<span class="co">##  $ crisprtranscript: chr  "TGGGGGGCCTTGGCGT" "GGGGGGCCTTGGCGTC" ...</span>
<span class="co">##  $ crisprextension : chr  "ACGCCAAGGCCCCCCACCACTGCCCCAGC" "GACGCCAAGGCCCCCCACCACTGCCCCAG" ...</span>
<span class="co">##  $ crisprextrange  : chr  "chr20:4699572-4699600:-" "chr20:4699573-4699601:-" ...</span>
<span class="co">##  $ off             : num  0 0 0 0</span>
<span class="co">##  $ off0            : num  0 0 0 0</span>
<span class="co">##  $ Doench2014      : num  0.00995 0.04497 ...</span>
<span class="co">##  $ nickrange       : chr  "chr20:4699632-4699651:-;chr20:4699633-4699652:-;chr20:4699664-4699683:-" "chr20:4699632-4699651:-;chr20:4699633-4699652:-;chr20:4699664-4699683:-" ...</span>
<span class="co">##  $ nickspacer      : chr  "TCACTGCCGAAATGTATGAT;GTCACTGCCGAAATGTATGA;GCATGTTTTCACGATAGTAA" "TCACTGCCGAAATGTATGAT;GTCACTGCCGAAATGTATGA;GCATGTTTTCACGATAGTAA" ...</span>
<span class="co">##  $ nickpam         : chr  "GGG;TGG;CGG" "GGG;TGG;CGG" ...</span>
<span class="co">##  $ nickoff         : chr  "0;0;0" "0;0;0" ...</span>
<span class="co">##  $ nickoff0        : chr  "0;0;0" "0;0;0" ...</span>
<span class="co">##  $ nickoff1        : chr  "0;0;0" "0;0;0" ...</span>
<span class="co">##  $ nickoff2        : chr  "0;0;0" "0;0;0" ...</span>
<span class="co">##  $ nickDoench2014  : chr  "0.14;0.32;0.52" "0.14;0.32;0.52" ...</span>
<span class="co">##  $ names           : chr  "PRNP_1" "PRNP_2" ...</span>
<span class="co">##  - attr(*, ".internal.selfref")=&lt;externalptr&gt;</span></pre></body></html></div>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>Anzalone, A.V., Randolph, P.B., Davis, J.R. et al. Search-and-replace genome editing without double-strand breaks or donor DNA. Nature 576, 149–157 (2019). <a href="https://doi.org/10.1038/s41586-019-1711-4" class="uri">https://doi.org/10.1038/s41586-019-1711-4</a></p>
<p>Doench et al. (2016). Optimized sgRNA design to maximize activity and minimize off-target effects of CRISPR-Cas9. Nature Biotechnology, 34(7), <a href="doi:10.1038/nbt.3437" class="uri">doi:10.1038/nbt.3437</a></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Aditya Bhagwat.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
